{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "_cell_guid": "3943a80b-df78-4fd7-8b05-7958300f86d5",
    "_execution_state": "idle",
    "_uuid": "23e8a136bbbd3f65e47c038e6ca25e3d3315472a"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "# This Python 3 environment comes with many helpful analytics libraries installed\n",
    "# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python\n",
    "# For example, here's several helpful packages to load in \n",
    "\n",
    "import numpy as np # linear algebra\n",
    "import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "import os\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import spacy\n",
    "import re\n",
    "from tqdm import tqdm\n",
    "CURRENT_DIR = os.getcwd()\n",
    "import time\n",
    "nlp = spacy.load('en')\n",
    "import string\n",
    "punc = set(string.punctuation)\n",
    "import pylab as pl\n",
    "from stemming.porter2 import stem\n",
    "from gensim import utils\n",
    "lem_dict = {} #for memoization\n",
    "# Input data files are available in the \"../input/\" directory.\n",
    "# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory\n",
    "\n",
    "# Any results you write to the current directory are saved as output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "_cell_guid": "0f057d77-c2d9-404a-b90a-ca9c7b73661a",
    "_execution_state": "idle",
    "_uuid": "fd150182db22059d694db55471c375d64146cb21"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading global stemming dictionary\n",
      "Processing and saving data...\n",
      "Skipping tokenizing, lowercasing and punc removal.\n",
      "Skipping stemming.\n",
      "Data saved.\n",
      "Training Dataframe shape: (3321, 7)\n",
      "Testing Dataframe shape: (5668, 6)\n"
     ]
    }
   ],
   "source": [
    "##Separate stem function that makes use of memoization to decrease time taken for stemming over 400000 words\n",
    "\n",
    "def load_stem_dict(force = False):\n",
    "    if not os.path.exists('global_stem_dict.pkl') or force ==True:\n",
    "        global_stem_dict = {}\n",
    "        global_stem_flag = False\n",
    "    else:\n",
    "        print(\"Loading global stemming dictionary\")\n",
    "        global_stem_dict = pickle.load(open('global_stem_dict.pkl',\"rb\"))\n",
    "        global_stem_flag = True\n",
    "    return global_stem_dict, global_stem_flag\n",
    "#global_stem_dict will hold words as keys and stemmed words as values\n",
    "#global_stem_flag is True if global_stem_dict was loaded from pickle file, False otherwise\n",
    "#this flag is so that we unnecessarily don't write into the pickle file again in the process_data function\n",
    "global_stem_dict,global_stem_flag = load_stem_dict(force = False)\n",
    "\n",
    "\n",
    "\n",
    "def stem2(sentence):\n",
    "    words = sentence.split()\n",
    "    for i in range(0,len(words)):\n",
    "        try:\n",
    "            words[i] = global_stem_dict[words[i]]\n",
    "        except KeyError:\n",
    "            s =  stem(words[i])\n",
    "            global_stem_dict[words[i]] = s\n",
    "            words[i] = s\n",
    "    return ' '.join(words)\n",
    "\n",
    "   \n",
    "\n",
    "\n",
    "def process_data(stemmed = False, tokenized = False, force = False):\n",
    "    #Load the global stemming dictionary\n",
    "    global global_stem_dict\n",
    "    global global_stem_flag\n",
    "    if not os.path.exists(\"processed_data.pkl\") or force == True:\n",
    "        ##Input the data\n",
    "        print(\"Processing and saving data...\")\n",
    "        train_variants = pd.read_csv(\"./training_variants\")\n",
    "        test_variants = pd.read_csv(\"./test_variants\")\n",
    "        train_temp_df = pd.read_csv(\"./training_text\",sep = \"\\|\\|\", engine = \"python\", \n",
    "                                 header = None, skiprows = 1,names = [\"ID\",\"Text\"] )\n",
    "        test_df = pd.read_csv(\"./test_text\",sep = \"\\|\\|\", engine = \"python\", \n",
    "                                 header = None, skiprows = 1,names = [\"ID\",\"Text\"] )\n",
    "\n",
    "\n",
    "        #Join the text and variants columns\n",
    "        train_temp_df = train_temp_df.join(train_variants.set_index(\"ID\"), on = \"ID\")\n",
    "        test_df = test_df.join(test_variants.set_index(\"ID\"), on = \"ID\")\n",
    "        \n",
    "        \n",
    "        train_temp_df[\"Text\"].astype(str)\n",
    "        test_df[\"Text\"].astype(str)\n",
    "        \n",
    "        if tokenized == True:\n",
    "            start = time.time()\n",
    "            print(\"Lowercasing and tokenizing now\")\n",
    "            train_temp_df[\"Text\"] = train_temp_df[\"Text\"].apply(lambda sen: ' '.join(list(utils.tokenize(sen.lower()))))\n",
    "            test_df[\"Text\"] = test_df[\"Text\"].apply(lambda sen: ' '.join(list(utils.tokenize(sen.lower()))))\n",
    "            print(\"Time taken for tokenization: \" + str(time.time() - start))\n",
    "        else:\n",
    "            print(\"Skipping tokenizing, lowercasing and punc removal.\")\n",
    "        \n",
    "        if stemmed == True:\n",
    "            print(\"Stemming now\")\n",
    "            start = time.time()\n",
    "            ##Stemming starts here\n",
    "            train_temp_df[\"Text\"] = train_temp_df[\"Text\"].apply(stem2)\n",
    "            test_df[\"Text\"] = test_df[\"Text\"].apply(stem2)\n",
    "            #Save into 'global_stem_dict.pkl' so that it can be loaded later by load_stem_dict() function\n",
    "            if global_stem_flag == True:\n",
    "                print(\"Stemmed words were loaded from existing pickle file. Skipping pickling.\")\n",
    "            else:\n",
    "                print(\"Global stem dictionary was reinitialized. Pickling...\")\n",
    "                pickle.dump(global_stem_dict,open(\"global_stem_dict.pkl\",\"wb\"), protocol=2)\n",
    "            print(\"Time taken for stemming operation: \" + str(time.time() - start))\n",
    "        else:\n",
    "            print(\"Skipping stemming.\")\n",
    "        \n",
    "        ## Concatenate the Gene and Variation data along with the Text field\n",
    "        train_temp_df['Text'] = train_temp_df.Text.str.cat(train_temp_df.Gene, sep=' ').str.cat(train_temp_df.Variation, sep=' ')\n",
    "        test_df['Text'] = test_df.Text.str.cat(test_df.Gene, sep=' ').str.cat(test_df.Variation, sep=' ')\n",
    "        #Add text counts as a feature\n",
    "        train_temp_df.loc[:,\"Text_Word_Count\"] = train_temp_df[\"Text\"].apply(lambda x: len(x.split()))\n",
    "        test_df.loc[:,\"Text_Word_Count\"] = test_df[\"Text\"].apply(lambda x: len(x.split()))\n",
    "        train_temp_df.loc[:,\"Text_Char_Count\"] = train_temp_df[\"Text\"].apply(lambda x: len(x))\n",
    "        test_df.loc[:,\"Text_Char_Count\"] = test_df[\"Text\"].apply(lambda x: len(x))\n",
    "        processed_data = {\"train_temp_df\": train_temp_df,\n",
    "                          \"test_df\": test_df\n",
    "                         }\n",
    "        pickle.dump(processed_data,open(\"processed_data.pkl\",'wb'), protocol = 2)\n",
    "        print(\"Data saved.\")\n",
    "\n",
    "    else:\n",
    "        print(\"Loading existing data\")\n",
    "        processed_data = pickle.load(open(\"processed_data.pkl\",\"rb\"))\n",
    "        train_temp_df = processed_data[\"train_temp_df\"]\n",
    "        test_df = processed_data[\"test_df\"]\n",
    "        print(\"Existing data loaded.\")\n",
    "    return train_temp_df, test_df\n",
    "\n",
    "train_temp_df, test_df = process_data(stemmed = False, tokenized = False, force = True)\n",
    "\n",
    "\n",
    "\n",
    "print(\"Training Dataframe shape: \" + str(train_temp_df.shape))\n",
    "print(\"Testing Dataframe shape: \" + str(test_df.shape))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import spacy\n",
    "nlp = spacy.load('en')\n",
    "doc = nlp(unicode(train_temp_df[\"Text\"][0],errors = \"ignore\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "ename": "LookupError",
     "evalue": "\n**********************************************************************\n  Resource u'corpora/stopwords' not found.  Please use the NLTK\n  Downloader to obtain the resource:  >>> nltk.download()\n  Searched in:\n    - '/home/nitin/nltk_data'\n    - '/usr/share/nltk_data'\n    - '/usr/local/share/nltk_data'\n    - '/usr/lib/nltk_data'\n    - '/usr/local/lib/nltk_data'\n**********************************************************************",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLookupError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-32-42d645faa019>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mnltk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcorpus\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mstopwords\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mstops\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstopwords\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwords\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'english'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0msentences\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdoc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msents\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mword\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mword\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msentences\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mword\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mstops\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#print(list((sen. for sen in sentences[0:10])))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nitin/anaconda2/envs/tensorflow/lib/python2.7/site-packages/nltk/corpus/util.pyc\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, attr)\u001b[0m\n\u001b[1;32m    114\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"LazyCorpusLoader object has no attribute '__bases__'\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 116\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__load\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    117\u001b[0m         \u001b[0;31m# This looks circular, but its not, since __load() changes our\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m         \u001b[0;31m# __class__ to something new:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nitin/anaconda2/envs/tensorflow/lib/python2.7/site-packages/nltk/corpus/util.pyc\u001b[0m in \u001b[0;36m__load\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     79\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mLookupError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mroot\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnltk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{}/{}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubdir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mzip_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 81\u001b[0;31m                 \u001b[0;32mexcept\u001b[0m \u001b[0mLookupError\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m         \u001b[0;31m# Load the corpus.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mLookupError\u001b[0m: \n**********************************************************************\n  Resource u'corpora/stopwords' not found.  Please use the NLTK\n  Downloader to obtain the resource:  >>> nltk.download()\n  Searched in:\n    - '/home/nitin/nltk_data'\n    - '/usr/share/nltk_data'\n    - '/usr/local/share/nltk_data'\n    - '/usr/lib/nltk_data'\n    - '/usr/local/lib/nltk_data'\n**********************************************************************"
     ]
    }
   ],
   "source": [
    "from nltk.corpus import stopwords\n",
    "stops = set(stopwords.words('english'))\n",
    "sentences = list(doc.sents)\n",
    "print([word for word in sentences[0] if not word in stops])\n",
    "#print(list((sen. for sen in sentences[0:10])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "_cell_guid": "4e373376-6a30-4af7-b50e-ceb3d8809510",
    "_execution_state": "idle",
    "_uuid": "ccb3f77f3de0722985169deba9416766c00d9c31"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Class\n",
      "1    454\n",
      "2    361\n",
      "3     71\n",
      "4    549\n",
      "5    194\n",
      "6    220\n",
      "7    762\n",
      "8     15\n",
      "9     30\n",
      "Name: Class, dtype: int64, Class\n",
      "1    114\n",
      "2     91\n",
      "3     18\n",
      "4    137\n",
      "5     48\n",
      "6     55\n",
      "7    191\n",
      "8      4\n",
      "9      7\n",
      "Name: Class, dtype: int64)\n",
      "RangeIndex(start=0, stop=2656, step=1)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\"\"\"\n",
    "#Tried this splitting method and found that classes 8 and 9 were not represented\n",
    "#in the validation set.\n",
    "from sklearn.model_selection import train_test_split\n",
    "#Split the training data into train and validation sets\n",
    "train_df, valid_df = train_test_split(train_temp_df, test_size = 0.2)\n",
    "\"\"\"\n",
    "#Lets use StratifiedShuffleSplit to represent each class equally in both train and validation sets\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "stratSplit = StratifiedShuffleSplit( n_splits = 3, test_size=0.2,\n",
    "                                    random_state=42).split(train_temp_df, train_temp_df[\"Class\"].tolist())\n",
    "\n",
    "ss = list(stratSplit)[0]\n",
    "\n",
    "\n",
    "train_df = train_temp_df.iloc[ss[0]].reset_index()\n",
    "valid_df = train_temp_df.iloc[ss[1]].reset_index()\n",
    "\n",
    "##Lets make sure that the classes are represented equalling \n",
    "tc = train_df.groupby(\"Class\")[\"Class\"].count()\n",
    "vc = valid_df.groupby(\"Class\")[\"Class\"].count()\n",
    "print(tc,vc)\n",
    "print(train_df[\"Text\"].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(135, 135)\n"
     ]
    }
   ],
   "source": [
    "from imblearn.under_sampling import NearMiss\n",
    "nm = NearMiss(random_state=42, version = 2)\n",
    "X_res, y_res = nm.fit_sample(train_feats, train_labels)\n",
    "print(len(X_res), len(y_res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "_cell_guid": "7ff7b382-fd90-44d1-98f2-ddea42d060b5",
    "_execution_state": "idle",
    "_uuid": "13200c13cb64f6db3e3aef16b3cd8bbca7d634fc",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading word stats.\n",
      "Total number of words: 31697301\n",
      "Total number of unique words: 488591\n",
      "Total number of unique words after lower case: 457437\n",
      "Total number of unique words after lower case and stemming: 439816\n",
      "Stats loaded from word_stats.pkl\n"
     ]
    }
   ],
   "source": [
    "from stemming.porter2 import stem\n",
    "\n",
    "def ngram_hash(sentence,n=3):\n",
    "    sentence = sentence.replace(' ','#')\n",
    "    return ' '.join(sentence[i:i+n] for i in range(0,len(sentence)-n))\n",
    "\n",
    "\n",
    "def find_word_stats(df,fname,force = False):\n",
    "    if not os.path.exists(fname + \".pkl\") or  force == True:\n",
    "        print(\"Calculating word stats.\")\n",
    "        #Let's see how many words are there\n",
    "        all_words = ' '.join(df[\"Text\"]).split()\n",
    "        unique_words = set(all_words)\n",
    "        lower_unique = set(' '.join(df[\"Text\"]).lower().split())\n",
    "        stemmed_lower_unique = set([stem(w) for w in lower_unique])\n",
    "        print(\"Total number of words: \" + str(len(all_words)))\n",
    "        print(\"Total number of unique words: \" + str(len(unique_words)))\n",
    "        print(\"Total number of unique words after lower case: \" + str(len(lower_unique)))\n",
    "        print(\"Total number of unique words after lower case and stemming: \" + str(len(stemmed_lower_unique)))\n",
    "        word_stats = {\n",
    "            \"len_all_words\": str(len(all_words)),\n",
    "            \"len_unique_words\": str(len(unique_words)),\n",
    "            \"len_lower_unique\": str(len(lower_unique)),\n",
    "            \"len_stemmed_lower_unique\": str(len(stemmed_lower_unique))\n",
    "        }\n",
    "        pickle.dump(word_stats,open(fname + \".pkl\",\"wb\"),protocol=2)\n",
    "        print(\"Stats stored in word_stats.pkl\")\n",
    "    else:\n",
    "        print(\"Loading word stats.\")\n",
    "        word_stats = pickle.load(open(fname + \".pkl\",\"rb\"))\n",
    "        len_all_words = word_stats[\"len_all_words\"]\n",
    "        len_unique_words = word_stats[\"len_unique_words\"]\n",
    "        len_lower_unique = word_stats[\"len_lower_unique\"]\n",
    "        len_stemmed_lower_unique = word_stats[\"len_stemmed_lower_unique\"]\n",
    "        print(\"Total number of words: \" + len_all_words)\n",
    "        print(\"Total number of unique words: \" + len_unique_words)\n",
    "        print(\"Total number of unique words after lower case: \" + len_lower_unique)\n",
    "        print(\"Total number of unique words after lower case and stemming: \" + len_stemmed_lower_unique)\n",
    "        print(\"Stats loaded from word_stats.pkl\")\n",
    "    return word_stats\n",
    "\n",
    "train_word_stats = find_word_stats(train_temp_df,\"train_temp_df_stats\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading existing model\n",
      "Word2Vec model and word vectors loaded...\n",
      "The most similar words to 'clinical' are: [('therapeutic', 0.608913779258728), ('diagnostic', 0.5986784100532532), ('clinicopathological', 0.593536913394928), ('clinicopathologic', 0.5916438102722168), ('randomised', 0.5887097120285034), ('pathologic', 0.5514856576919556), ('prospective', 0.5512081384658813), ('prognostic', 0.5504527688026428), ('safety', 0.5188233256340027), ('ongoing', 0.5186888575553894)]\n"
     ]
    }
   ],
   "source": [
    "##Let's add word2vec features with gensim\n",
    "## We will train it with the only processing used being the lowercase filter.\n",
    "from gensim.models import Word2Vec\n",
    "from gensim.models.keyedvectors import KeyedVectors\n",
    "\n",
    "def build_w2v(text_array, force = False):\n",
    "    if not os.path.exists(\"./w2v_model\") or force == True:\n",
    "        print(\"Model does not exist. Training model...\")\n",
    "        w2v_model = Word2Vec(train_temp_df[\"Text\"].apply(lambda x: x.lower().split()), \n",
    "                             size=100, window=5, min_count=5, workers=4)\n",
    "        w2v_model.save(\"w2v_model\")\n",
    "        word_vectors = w2v_model.wv\n",
    "        word_vectors.save(\"word_vectors\")\n",
    "        print(\"Model and Word Vectors trained and saved...\")\n",
    "        print(\"The most similar words to 'clinical' are: \" + str(word_vectors.most_similar('clinical')))\n",
    "    else:\n",
    "        print(\"Loading existing model\")\n",
    "        w2v_model = Word2Vec.load(\"w2v_model\")\n",
    "        word_vectors = KeyedVectors.load(\"word_vectors\")\n",
    "        print(\"Word2Vec model and word vectors loaded...\")\n",
    "        print(\"The most similar words to 'clinical' are: \" + str(word_vectors.most_similar('clinical')))\n",
    "    return w2v_model, word_vectors\n",
    "\n",
    "w2v_model, word_vectors = build_w2v(train_temp_df[\"Text\"])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading existing model...\n",
      "Doc2Vec model loaded...\n",
      "Time taken: 6.62736701965\n",
      "Loading existing model...\n",
      "Doc2Vec model loaded...\n",
      "Time taken: 3.23418712616\n",
      "Loading existing model...\n",
      "Doc2Vec model loaded...\n",
      "Time taken: 8.49217796326\n"
     ]
    }
   ],
   "source": [
    "##Let's perform doc2vec\n",
    "from gensim.models import Doc2Vec\n",
    "from gensim.models.doc2vec import LabeledSentence\n",
    "from gensim import utils\n",
    "\n",
    "class LabeledLineSentence(object):\n",
    "    def __init__(self, df):\n",
    "        self.text_array = df[\"Text\"].tolist()\n",
    "        self.labels = df[\"Class\"].tolist()\n",
    "    def __iter__(self):\n",
    "        for i,line in tqdm(enumerate(self.text_array)):\n",
    "            yield LabeledSentence(words=utils.to_unicode(line.lower()).split(), tags=[\"Class\" + str(self.labels[i])])\n",
    "\n",
    "def build_d2v(text_array, fname, force = False):\n",
    "    start = time.time()\n",
    "    if not os.path.exists(fname) or force == True:\n",
    "        print(\"Training Doc2Vec model...\")\n",
    "        lb = LabeledLineSentence(text_array)\n",
    "        d2v_model = Doc2Vec(lb, size=100, window=8, min_count=1, workers=4)\n",
    "        d2v_model.save(fname)\n",
    "        print(\"Model saved.\")\n",
    "    else:\n",
    "        print(\"Loading existing model...\")\n",
    "        d2v_model = Doc2Vec.load(fname)  # you can continue training with the loaded model!\n",
    "        print(\"Doc2Vec model loaded...\")\n",
    "    print(\"Time taken: \" + str(time.time() - start))\n",
    "    return d2v_model\n",
    "\n",
    "d2v_model_train = build_d2v(train_df,\"d2v_model_train\", force = False)\n",
    "d2v_model_valid = build_d2v(valid_df[\"Text\"].tolist(),\"d2v_model_valid\", force = False)\n",
    "d2v_model_test = build_d2v(test_df[\"Text\"].tolist(),\"d2v_model_test\", force = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "_cell_guid": "369241a5-b432-4a31-a5ac-87f3642c6b62",
    "_execution_state": "idle",
    "_uuid": "ce30913f00f824fa84e971a3fc84f78c19b13291",
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "##Create training and validation labels\n",
    "train_labels = train_df[\"Class\"].tolist()\n",
    "valid_labels = valid_df[\"Class\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "_cell_guid": "d219b56f-c2ad-438c-86a3-59408f0e0b76",
    "_execution_state": "idle",
    "_uuid": "6236f4820e64475e8883b796d442eb3f796ffe48"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created and fit the TFIDFvectorizer with text data.\n",
      "Saved TFIDF Vectorizer in a pickle file.\n",
      "Transformed train, valid and test text arrays.\n",
      "Storing TFIDF features into a pickle file.\n",
      "Finished vectorizing the data\n",
      "Train Tfidf features shape: (2656, 8000)\n",
      "Validation Tfidf features shape: (665, 8000)\n",
      "Test Tfidf features shape: (5668, 8000)\n",
      "TfidfVectorizer(analyzer=u'word', binary=False, decode_error=u'strict',\n",
      "        dtype=<type 'numpy.int64'>, encoding=u'utf-8', input=u'content',\n",
      "        lowercase=True, max_df=1.0, max_features=8000, min_df=5,\n",
      "        ngram_range=(1, 3), norm=u'l2', preprocessor=None, smooth_idf=True,\n",
      "        stop_words='english', strip_accents='unicode', sublinear_tf=True,\n",
      "        token_pattern='\\\\w+', tokenizer=None, use_idf=True,\n",
      "        vocabulary=None)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer\n",
    "\n",
    "\n",
    "if os.path.exists(os.path.join(CURRENT_DIR, 'PM_data.pkl')):\n",
    "    data_to_store = pickle.load(open('PM_data.pkl','rb'))\n",
    "    train_tfidf_feats = data_to_store[\"train_tfidf_feats\"]\n",
    "    valid_tfidf_feats = data_to_store[\"valid_tfidf_feats\"]\n",
    "    test_tfidf_feats = data_to_store[\"test_tfidf_feats\"]\n",
    "    tfidf = pickle.load(open(\"TFIDF_Vectorizer.pkl\",\"rb\"))\n",
    "    print(\"Loaded existing TFIDF data\")\n",
    "else:\n",
    "    tfidf = TfidfVectorizer(lowercase = True, use_idf = True, strip_accents = \"unicode\",\n",
    "                            min_df = 5, max_features = 16000, token_pattern = r\"\\w+\",\n",
    "                            smooth_idf=True, sublinear_tf=True,ngram_range=(1,3),\n",
    "                            stop_words = 'english').fit(train_temp_df[\"Text\"])\n",
    "    \n",
    "    print(\"Created and fit the TFIDFvectorizer with text data.\")\n",
    "    \n",
    "    pickle.dump(tfidf,open(\"TFIDF_Vectorizer.pkl\",\"wb\"),protocol=2)\n",
    "    print(\"Saved TFIDF Vectorizer in a pickle file.\")\n",
    "    train_tfidf_feats = tfidf.transform(train_df[\"Text\"])\n",
    "    valid_tfidf_feats = tfidf.transform(valid_df[\"Text\"])\n",
    "    test_tfidf_feats = tfidf.transform(test_df[\"Text\"])\n",
    "    print(\"Transformed train, valid and test text arrays.\")\n",
    "\n",
    "    #Convert each tfidf array into a numpy array so that we can append additional features to it later.\n",
    "    train_tfidf_feats = train_tfidf_feats.toarray()\n",
    "    valid_tfidf_feats = valid_tfidf_feats.toarray()\n",
    "    test_tfidf_feats = test_tfidf_feats.toarray()\n",
    "    print(\"Storing TFIDF features into a pickle file.\")\n",
    "    data_to_store = {\n",
    "            \"train_tfidf_feats\": train_tfidf_feats,\n",
    "            \"valid_tfidf_feats\": valid_tfidf_feats,\n",
    "            \"test_tfidf_feats\": test_tfidf_feats}\n",
    "    pickle.dump(data_to_store, open(\"PM_data.pkl\",\"wb\"), protocol = 2)\n",
    "    \n",
    "    print(\"Finished vectorizing the data\")\n",
    "print(\"Train Tfidf features shape: \" + str(train_tfidf_feats.shape))\n",
    "print(\"Validation Tfidf features shape: \" + str(valid_tfidf_feats.shape))\n",
    "print(\"Test Tfidf features shape: \" + str(test_tfidf_feats.shape))\n",
    "print(tfidf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading train_df_doc_vectors doc vectors\n",
      "Time taken 0.13977599144\n",
      "Loading valid_df_doc_vectors doc vectors\n",
      "Time taken 0.183408021927\n",
      "Loading test_df_doc_vectors doc vectors\n",
      "Time taken 0.0234789848328\n"
     ]
    }
   ],
   "source": [
    "def form_doc_vector(text_array,fname, force = False):\n",
    "    start = time.time()\n",
    "    if not os.path.exists(fname + \".pkl\") or force == True:\n",
    "        print(\"Creating \" + fname + \" doc vectors.\")\n",
    "        \n",
    "        temp_array = np.zeros([len(text_array),100]).astype(np.float32)\n",
    "        print(\"Shape of array created: \" + str(temp_array.shape))\n",
    "        for i in tqdm(range(0,len(text_array))):\n",
    "            for word in text_array[i].split():\n",
    "                try:\n",
    "                    temp_array[i] += word_vectors[word]\n",
    "                except KeyError:\n",
    "                    pass\n",
    "            temp_array[i] = temp_array[i]/temp_array.shape[0]\n",
    "        pickle.dump(temp_array, open(fname + \".pkl\", \"wb\"),protocol=2)\n",
    "    else:\n",
    "        print(\"Loading \" + fname + \" doc vectors\")\n",
    "        temp_array = pickle.load(open(fname + \".pkl\", \"rb\"))\n",
    "    print(\"Time taken \" + str(time.time() - start))\n",
    "    return temp_array\n",
    "    \n",
    "train_df_doc_vectors = form_doc_vector(train_df[\"Text\"],\"train_df_doc_vectors\")\n",
    "valid_df_doc_vectors = form_doc_vector(valid_df[\"Text\"],\"valid_df_doc_vectors\")\n",
    "test_df_doc_vectors = form_doc_vector(test_df[\"Text\"],\"test_df_doc_vectors\")\n",
    "#train_df[\"Text\"] = train_df[\"Text\"].reset_index()\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "_cell_guid": "c7688957-a1be-4647-8df4-9be729d4b153",
    "_execution_state": "idle",
    "_uuid": "96c066f8402eb131712bf74acdad4aa6767afee2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of Train TFIDF features: (2656, 8000)\n",
      "Shape of Validation TFIDF features: (665, 8000)\n",
      "Shape of Test TFIDF features: (5668, 8000)\n",
      "Word Counts shape: (2656, 1)\n",
      "8002\n",
      "Train data shape after feature engineering: (2656, 8002)\n",
      "Validation data shape after feature engineering: (665, 8002)\n",
      "Test data shape after feature engineering: (5668, 8002)\n"
     ]
    }
   ],
   "source": [
    "#Let's add text count as an additional feature after mean normalization.\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler()\n",
    "train_word_counts = scaler.fit_transform(train_df[\"Text_Word_Count\"].values.reshape(-1,1))\n",
    "valid_word_counts = scaler.fit_transform(valid_df[\"Text_Word_Count\"].values.reshape(-1,1))\n",
    "test_word_counts = scaler.fit_transform(test_df[\"Text_Word_Count\"].values.reshape(-1,1))\n",
    "train_char_counts = scaler.fit_transform(train_df[\"Text_Char_Count\"].values.reshape(-1,1))\n",
    "valid_char_counts = scaler.fit_transform(valid_df[\"Text_Char_Count\"].values.reshape(-1,1))\n",
    "test_char_counts = scaler.fit_transform(test_df[\"Text_Char_Count\"].values.reshape(-1,1))\n",
    "\n",
    "\"\"\"\n",
    "#doc2vec features - 100 in size\n",
    "train_temp , valid_temp, test_temp = [], [], []\n",
    "for i in range(0,len(train_df)):\n",
    "    train_temp.append(d2v_model_train[str(i)])\n",
    "for i in range(0,len(valid_df)):\n",
    "    valid_temp.append(d2v_model_valid[str(i)])\n",
    "for i in range(0,len(test_df)):\n",
    "    test_temp.append(d2v_model_test[str(i)])\n",
    "\"\"\"  \n",
    "\n",
    "#train_gene_feats = np.array(pd.get_dummies(train_df[\"Gene\"]))\n",
    "#valid_gene_feats = np.array(pd.get_dummies(valid_df[\"Gene\"]))\n",
    "#test_gene_feats = np.array(pd.get_dummies(test_df[\"Gene\"]))\n",
    "#train_variation_feats = np.array(pd.get_dummies(train_df[\"Variation\"]))\n",
    "#valid_variation_feats = np.array(pd.get_dummies(valid_df[\"Variation\"]))\n",
    "#test_variation_feats = np.array(pd.get_dummies(test_df[\"Variation\"]))\n",
    "\n",
    "print(\"Shape of Train TFIDF features: \" + str(train_tfidf_feats.shape))\n",
    "print(\"Shape of Validation TFIDF features: \" + str(valid_tfidf_feats.shape))\n",
    "print(\"Shape of Test TFIDF features: \" + str(test_tfidf_feats.shape))\n",
    "print(\"Word Counts shape: \" + str(train_word_counts.shape))\n",
    "\n",
    "#Append this to the train and text features array\n",
    "train_feats = np.hstack([train_tfidf_feats,train_word_counts, train_char_counts])\n",
    "valid_feats = np.hstack([valid_tfidf_feats,valid_word_counts, valid_char_counts])\n",
    "test_feats = np.hstack([test_tfidf_feats,test_word_counts, test_char_counts])\n",
    "train_labels = train_df[\"Class\"].tolist()\n",
    "valid_labels = valid_df[\"Class\"].tolist()\n",
    "##Let's shuffle the data\n",
    "\n",
    "np.random.seed = 42\n",
    "\n",
    "temp_t = zip(train_feats, train_labels)\n",
    "np.random.shuffle(temp_t)\n",
    "train_feats, train_labels = zip(*temp_t)\n",
    "train_labels = list(train_labels)\n",
    "train_feats = np.array(train_feats)\n",
    "\n",
    "temp_v = zip(valid_feats, valid_labels)\n",
    "np.random.shuffle(temp_v)\n",
    "valid_feats, valid_labels = zip(*temp_v)\n",
    "valid_labels = list(valid_labels)\n",
    "valid_feats = np.array(valid_feats)\n",
    "\n",
    "feature_names = tfidf.get_feature_names() + [\"Text Word Counts\", \"Text Char Counts\"]\n",
    "print(len(feature_names))\n",
    "\n",
    "print(\"Train data shape after feature engineering: \" + str(train_feats.shape))\n",
    "print(\"Validation data shape after feature engineering: \" + str(valid_feats.shape))\n",
    "print(\"Test data shape after feature engineering: \" + str(test_feats.shape))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[u'axl 1.092',\n",
       " u'ac uk 1.101',\n",
       " u'brca brct domain 1.105',\n",
       " u'akt 1.144',\n",
       " u'1 1.184',\n",
       " u'allogeneic 1.213',\n",
       " u'brca dna 1.223',\n",
       " u'denaturing 1.241',\n",
       " u'deletion exon 1.315',\n",
       " u'conjunction 1.324',\n",
       " u'bonds 1.363',\n",
       " u'acc 1.381',\n",
       " u'catalytic site 1.381',\n",
       " u'consideration 1.39',\n",
       " u'characterized 1.47',\n",
       " u'bcl 1.474',\n",
       " u'bay 1.589',\n",
       " u'cancer associated mutations 1.834',\n",
       " u'deleterious neutral 1.998',\n",
       " u'deleterious variants 2.012']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def plot_coefficients(classifier, vectorizer, top_features=20):\n",
    "    tfidf_vocab = dict((v,k) for k,v in vectorizer.vocabulary_.iteritems())\n",
    "    coef = classifier.coef_[0].ravel()\n",
    "    top_positive_coefficients = np.argsort(coef)[-top_features:]\n",
    "    top_negative_coefficients = np.argsort(coef)[:top_features]\n",
    "    return top_positive_coefficients, top_negative_coefficients\n",
    "    top_neg_feature_names = [tfidf_vocab[k] for k in top_negative_coefficients]\n",
    "    top_pos_feature_names = [tfidf_vocab[k] for k in top_positive_coefficients]\n",
    "    feature_names = top_neg_feature_names + top_pos_feature_names\n",
    "    top_coefficients = np.hstack([top_negative_coefficients, top_positive_coefficients])\n",
    "    # create plot\n",
    "    plt.figure(figsize=(15, 5))\n",
    "    colors = ['red' if c < 0 else 'blue' for c in coef[top_coefficients]]\n",
    "    plt.bar(np.arange(2 * top_features), coef[top_coefficients], color=colors)\n",
    "    #feature_names = np.array(feature_names)\n",
    "    \n",
    "    plt.xticks(np.arange(top_features), feature_names, rotation=60, ha='right')\n",
    "    plt.show()\n",
    "    return top_positive_coefficients, top_negative_coefficients\n",
    "\n",
    "def imp_features(clf, feature_names, topn =20):\n",
    "    pass\n",
    "[tfidf.get_feature_names()[k] + ' ' + str(round(svm.coef_[0][k],3)) for k in np.argsort(svm.coef_[0])[-20:]]\n",
    "#[tfidf.get_feature_names()[k] + ' ' + str(round(svm.coef_[0][k],3)) for k in np.argsort(svm.coef_[0])[:20]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "_cell_guid": "9ed5fb1b-1b6a-4185-a5dc-b8524833a0a0",
    "_execution_state": "idle",
    "_uuid": "ba305d884c02ea5ff7718e50fa6dfd532ea64d0e"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding 1000 best features using chi2 test\n",
      "[LibLinear]LinearSVC(C=1, class_weight=None, dual=True, fit_intercept=True,\n",
      "     intercept_scaling=1, loss='squared_hinge', max_iter=1000,\n",
      "     multi_class='ovr', penalty='l2', random_state=42, tol=0.0001,\n",
      "     verbose=True)\n",
      "[LibLinear][LibLinear][LibLinear][LibLinear][LibLinear]Cross Validation scores are: [ 0.63108614  0.63909774  0.62711864  0.63088512  0.63825758]\n",
      "Validation Accuracy: 60\n",
      "[[ 71   0   0  24  10   1   8   0   0]\n",
      " [  5  29   0   1   4   0  52   0   0]\n",
      " [  2   1   2   2   5   0   6   0   0]\n",
      " [ 35   0   1  84   2   3  12   0   0]\n",
      " [  9   1   0   6  15   2  15   0   0]\n",
      " [  8   1   1   5   2  26  12   0   0]\n",
      " [  2  13   0   0   3   0 173   0   0]\n",
      " [  0   0   0   1   0   0   2   0   1]\n",
      " [  0   1   0   1   0   0   0   0   5]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"\\nfig = plt.figure()\\nax = fig.add_subplot(111)\\ncax = ax.matshow(svm_conf_mat)\\npl.title('Confusion matrix of the classifier')\\nfig.colorbar(cax)\\nax.set_xticklabels([''] + labels)\\nax.set_yticklabels([''] + labels)\\npl.xlabel('Predicted')\\npl.ylabel('True')\\npl.show()\\n\""
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "def accuracy(predictions, actual):\n",
    "    temp_sum = 0\n",
    "    for i in range(len(predictions)):\n",
    "        if predictions[i] == actual[i]:\n",
    "            temp_sum += 1\n",
    "    return 100 * temp_sum/len(predictions) \n",
    "\n",
    "def make_submission(test_df, predictions):\n",
    "    classes = \"class1,class2,class3,class4,class5,class6,class7,class8,class9\".split(',')\n",
    "    print(classes)\n",
    "    a = pd.get_dummies(predictions)\n",
    "    a.head()\n",
    "    a.columns = classes\n",
    "    op_df = pd.concat([test_df[\"ID\"],a],axis =1)\n",
    "    print(op_df.head())\n",
    "    print(op_df.columns)\n",
    "    op_df.to_csv('submission.csv', index=False)\n",
    "    print(\"Submission file created.\")\n",
    "\n",
    "##We'll try to select K Best features using chi_square test\n",
    "from sklearn.feature_selection import SelectKBest, chi2\n",
    "\n",
    "print(\"Finding 1000 best features using chi2 test\")\n",
    "\n",
    "all_feats = np.vstack([train_feats,valid_feats])\n",
    "all_labels = train_labels + valid_labels\n",
    "\n",
    "ch2 = SelectKBest(chi2, k=2000)\n",
    "train_feats_chi2 = ch2.fit_transform(train_feats, train_labels)\n",
    "valid_feats_chi2 = ch2.transform(valid_feats)\n",
    "test_feats_chi2 = ch2.transform(test_feats)\n",
    "all_feats_chi2 = np.vstack([train_feats_chi2,valid_feats_chi2])\n",
    "\n",
    "\n",
    "from sklearn.svm import LinearSVC\n",
    "from sklearn import svm\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "    \n",
    "\n",
    "##Get all the data and merge it.\n",
    "#all_feats = np.vstack([train_tfidf_feats,valid_tfidf_feats])\n",
    "#all_labels = pd.concat([train_df,valid_df])\n",
    "#Base SVM with 5000 TFIDF features is giving 64% accuracy on validation set.\n",
    "#Base SVM with 5000 TFIDF features + 2 count features is giving 64% accuracy on validation set.\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "#Grid search returned C=1 and kernel = 'Linear'\n",
    "parameters = {'kernel':('linear', 'rbf'), 'C':[0.1,1, 10,100]}\n",
    "svm = svm.SVC(random_state = 42, verbose = 2)\n",
    "svm = GridSearchCV(svm, parameters).fit(train_feats,train_labels)\n",
    "\n",
    "\"\"\"\n",
    "svm = LinearSVC(C=1, verbose = True, random_state = 42).fit(train_feats_chi2,train_labels)\n",
    "print(svm)\n",
    "scores = cross_val_score(svm,train_feats_chi2, train_labels,  cv=5)\n",
    "print(\"Cross Validation scores are: \" + str(scores))\n",
    "\n",
    "predictions = svm.predict(valid_feats_chi2)\n",
    "test_preds = svm.predict(test_feats_chi2)\n",
    "#make_submission(test_df, test_preds)\n",
    "\n",
    "acc = accuracy(predictions, np.array(valid_labels))\n",
    "print(\"Validation Accuracy: \" + str(acc))\n",
    "\n",
    "\n",
    "#Print F1 score\n",
    "from sklearn.metrics import f1_score\n",
    "f1_score(valid_labels, predictions, average = \"weighted\")\n",
    "\n",
    "#Plot confusion matrix\n",
    "svm_conf_mat = confusion_matrix(valid_labels, predictions, labels = range(1,10))\n",
    "print(svm_conf_mat)\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "cax = ax.matshow(svm_conf_mat)\n",
    "pl.title('Confusion matrix of the classifier')\n",
    "fig.colorbar(cax)\n",
    "ax.set_xticklabels([''] + labels)\n",
    "ax.set_yticklabels([''] + labels)\n",
    "pl.xlabel('Predicted')\n",
    "pl.ylabel('True')\n",
    "pl.show()\n",
    "\"\"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation Accuracy: 50\n",
      "('Cross Validation Scores: ', array([ 0.53370787,  0.53571429,  0.52165725,  0.54613936,  0.56060606]))\n",
      "[[ 56   0   0  18   9   4  27   0   0]\n",
      " [  2   6   0   1   0   0  82   0   0]\n",
      " [  1   0   0   1   7   0   9   0   0]\n",
      " [ 43   0   0  54   4   3  33   0   0]\n",
      " [  9   0   0   6  13   3  17   0   0]\n",
      " [ 12   0   0   1   3  19  20   0   0]\n",
      " [  2   1   0   3   0   0 185   0   0]\n",
      " [  0   0   0   1   0   0   3   0   0]\n",
      " [  3   0   0   1   0   0   3   0   0]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUkAAAEQCAYAAAA9C4aXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XmYZFV9//H3ZzaGVZABwr4oYIBEwPkRIoHgDgRR80QF\nFcENTTQ/iEajJo8QTfIjCW6JCQkEIgqiCGIMooAmSDSAzsDI4qAsggwzMjAswgwMM92f3x/nNtQ0\nXdXVPfdWV/V8Xs9zn66699Y5p6q7v3XPcs+RbSIiYmwzproAERH9LEEyIqKDBMmIiA4SJCMiOkiQ\njIjoIEEyIqKDBMlRJG0s6T8lPSrpq+uRzpslXVln2aaKpEMl/bSBdCf8WUu6WtI76y7LqDxOlPT9\nBtP/lqQTWp7/laQHJf1S0i6SHpc0s6n8Y2JmTXUBJkvSm4D3Ay8AHgMWAX9te33/uP8A2A7Y2vba\nySZi+wLggvUsS+MkGdjT9h3tzrH9P8DeDWTf8bOWdBrwfNtvaSDvKWP7yJHHknYGPgDsant5tXuz\nKSlYjGkgryQlvR/4DPA3lH+yXYB/Bl5TQ/K7Aj9bnwA5nUhq8os0n3X5DFa0BMhJa/h3teGyPVAb\n8BzgceD1Hc7ZiBJEl1bbZ4CNqmOHA0so397LgWXA26pjfwk8Bayp8ngHcBpwfkvauwEGZlXPTwTu\nolzN/hx4c8v+77e87sXAj4BHq58vbjl2NfAJ4AdVOlcC89q8t5Hyf6il/K8FjgJ+BjwEfLTl/IOA\na4FHqnM/B8ypjl1TvZeV1ft9Y0v6fwb8EvjiyL7qNc+r8jiwer4D8CBweJvy/nr1/h4BbgWOafdZ\nj3rdEaOO/7ibzwo4GPjfKr8ftytXde7OwNeAB4AVwOfa/O4+C9wL/ApYCBw66vNdUB27H/hUtX8u\ncH6V7iPV73y7lvfwTuDlwBPAcPUeP8+z/76eA5xT/e7uA/4KmNlSzh8An65+J3811f+f03Gb8gJM\nuMDln2ftyB9Rm3M+DlwHbAtsU/3TfKI6dnj1+o8DsynBZRWwVXX8NNYNiqOfP/1HDGxa/XPsXR3b\nHti3evz0PxrwXOBh4PjqdcdVz7eujl8N3AnsBWxcPT+9zXsbKf/HqvK/q/on/xKwObAv8CSwR3X+\niyiBY1ZV9sXAKS3pmVKlHZ3+31K+bDamJUhW57yrSmcT4ArgjDZlnQ3cAXwUmAO8lBLY9h7rsx3j\n9c863umzAnakBKWjKLWkV1TPtxkj7ZmUIPrp6vc4F/id0b+76vlbgK2rz/ADlC+PudWxa4Hjq8eb\nAQdXj98N/Gf1Gc2sfg9btLyHd7Z83q2f7W6sGyS/DvxrVcZtgR8C724p51rgj6uybTzV/5/TcRvE\n6vbWwIPuXEV7M/Bx28ttP0C5ajm+5fia6vga25dTvsUn2+Y2DOwnaWPby2zfOsY5vwfcbvuLttfa\nvhC4DXh1yzn/bvtntp8ALgL275DnGkr76xrgy8A84LO2H6vyvxX4TQDbC21fV+V7N+Uf7ne7eE+n\n2l5dlWcdts8Gbgeup3wx/HmbdA6mBI7TbT9l+7+AyyhfEuuj3Wf1FuBy25fbHrZ9FeUq76gx0jiI\nchX8QdsrbT/pNu3Zts+3vaL6DD9J+fIY+XtZAzxf0jzbj9u+rmX/1pQvoKHq9/CribxJSdsBR1K+\n1Fa6VMk/DRzbctpS2/9Yle1Zv6tYf4MYJFcA88Zpf9kBuKfl+T3VvqfTGBVkVzGJxnLbKylV1PcA\nyyR9U9ILuijPSJl2bHn+ywmUZ4XtoerxyD/G/S3Hnxh5vaS9JF1W9Zz+itKOO69D2gAP2H5ynHPO\nBvYD/tH26jbn7ADca3u4Zd/o9z0Z7T6rXYHXS3pkZAN+hxLIR9sZuGecL1sAJH1A0uKqF/4RShV4\n5DN8B+Wq9jZJP5J0dLX/i5Sr7C9LWirp7yTNnuD73JVyNb6s5f38K+WKcsS9E0wzJmgQg+S1lOrk\nazucs5TyBzZil2rfZKykVJlG/FrrQdtX2H4F5R/xNkrwGK88I2W6b5JlmogzKeXa0/YWlKqvxnlN\nx6mhJG1Gaec9BzhN0nPbnLoU2FlS69/ZRN73RKeouhf4ou0tW7ZNbZ/e5txdxuvskHQopX32DZQm\nmS0p7coCsH277eMogetvgYslbVrVUv7S9j6U9uijgbdO4v2sprS5jryfLWzv23JOpvFq2MAFSduP\nUtrj/knSayVtImm2pCMl/V112oXAX0jaRtK86vzzJ5nlIuCwavzac4CPjByQtJ2kYyRtSvljfhwY\nGiONy4G9JL1J0ixJbwT2oVQ9m7Y5pd308eoq9w9HHb8f2GOCaX4WWGj7ncA3gX9pc971lC+ZD1W/\no8MpTQxf7jKf+4HdRgXZTs4HXi3pVZJmSpor6XBJO41x7g8pnSGnS9q0OveQMc7bnNLu9wAwS9LH\ngC1GDkp6i6RtqqvlR6rdQ5JeIuk3qvGOv6JUv8f622jL9jJKx9QnJW0haYak50kar7kkajRwQRLA\n9qcoYyT/gvLHey/wPkojN5QewAXATcDNwA3VvsnkdRXwlSqthawb2GZQGvKXUnoXfxf4ozHSWEG5\nkvgApbngQ8DRth+cTJkm6E+BN1E6TM6mvJdWpwHnVdW5N4yXmKTXUDrP3lPtej9woKQ3jz7X9lPA\nMZR2tQcpw7Teavu2Lss+MsB8haQbxjvZ9r2UYWAf5Zm/iw8yxt951VzxauD5wC8oPfpvHCPZK4Bv\nUUYO3EOpxbRWcY8AbpX0OOXL49iqqeLXgIspAXIx8D0m90X9Vkqn108onX0XM3bzQTRE9uBdrUs6\nlxJ0ltver+a0dwa+QPkjHwbOsv3ZmvOYSxl+sxGlV/Ji26fWnMdMyhfFfbaPHu/8SaR/NyXwDgFr\nbc+vOf0tgX+jtHsaeLvta2tKe2/W/bLYA/iY7c/UkX6Vx59QhvmY8kX9ti7aeSeS/smUUQYCzq6z\n7DHKVHatT3YDDgMOBG5pIO3teWYM4OaUK4h9as5DwGbV49mUaunBNefxfsqwoMsa+h3cTZuxnDWl\nfx7PDJOZA2zZUD4zKR1Bu9aY5o6UMbMbV88vAk6sMf39gFsobeWzgO9Q2pwb+V1s6NugVrevoVRv\nm0h7me0bqsePUapK69sbOzoP2368ejq72mq7pK/a4H6PciU2cCRtQfkiPAdKtd32I51fNWkvA+60\nPXr0wfqaBWxcdQxtwuQ7Dsfy68B1tle59M5/D3hdjelHi4EMkr0iaTfgAMqVXt1pz5S0iHLXzFW2\n68zjM5R2z+HxTlwPBq6UtFDSSTWnvQelTfHfJd0o6d+qzrEmHEvp6KuN7fuAMyhtncuAR23XOdnJ\nLZTOxK0lbUIZB7pzjelHiwTJNqphLpdQBvJOaBBwN1wGGO8P7AQcJKmWttVqnN5y2wvrSK+DQ2wf\nSOmUea+kw2pMexalOeVM2wdQesg/XGP6AEiaQ+lYmvRsT23S3YrSgbQ7ZazoppJqm6TD9mLKcKOr\ngG9T7hzakO9/b1SC5BiqQb+XABfY/lqTeVXVyKspvaR1OAQ4pupY+TLwUkmTHf7Ulu2l1c/lwKWU\nO1jqsoRyq97I1fXFlKBZtyOBG2zfP+6ZE/Ny4Oe2H3C5K+prlLGStbF9ju0DbR9GaXq6vc704xkJ\nkqNIEqUtbLHLUKMm8tim6r1F0saUf6puh8V0ZPsjtneyvRulKvlfrnmqsWpc4eYjj4FXUqqAtbD9\nS+DeqhcaSrvhT+pKv8Vx1FzVrvwCOLgawytK+RfXmYGkbaufuwC/TzPvIxjQ+SQlXUiZGGCepCWU\n+4zPqSn5Qyj3ed9ctRlCmVXn8prSh9KDfl41TGcGcJHtXgwsr8t2wKXl/59ZwJdsf7vmPP4YuKCq\nEt8FvK3OxKu2vFdQJqKole3rJV1MGZ+7FrgROKvmbC6RtDVlkPp7bT9cc/pRGchxkhERvZLqdkRE\nBwmSEREdJEhGRHSQIBkR0cFAB8kG7vRI+htQ+r3IY9DTjwEPkkDTfyBJf3qn34s8Bj39Dd6gB8mI\niEb11TjJWXM39UabtVsJ4NnWPrmSWXO7n/dg5oqVEyrPGlYzm40m9Jqm09ecOV2f+9TwKubM2GT8\nE1sNdT959lN+kjmaO6HkPdz9nBtr/CSzJ5g+G0/s81yzdhWzZ03wM1rV/bSQ/fY39CQrecqrx1u+\no1GvesmmXvFQd39nC29afYXtum7ZnZS+uuNmo82eyz5H/0lj6W/5hVrmbJ1Ss3YavVROvfxI7XN5\nrGN41apG0+cFz282fcA3jrUg5mC43t+d6iKw4qEhfnjFLl2dO3P728dbtK5xfRUkI2L6MzDc6Cx+\n9UqQjIieMmaNJ7Qm2pRKkIyInsuVZEREG8YM9VGH8XgSJCOi54brW9KpcQmSEdFTBoYSJCMi2huk\nK8nG7riRdK6k5ZJqm9Y/IgafgTV2V9t4xoozkr4iaVG13T2ywoCk3SQ90XLsX7opb5NXkp8HPgd8\nocE8ImLAGNdZ3f48o+KM7TeOPJb0SeDRlvPvrFYp7VpjQdL2NdW61RERzzAM1RQjO8WZahG2NwAv\nXZ88pnyCC0knSVogacHaJyd2b3VEDJ5yx013G2WxvwUt20RmPToUuN9263K7u0u6UdL3JB3aTSJT\n3nFj+yyqleQ2nbfz4LTmRsQkiSG6nmPjQdvzJ5nR6CWDlwG72F4h6UXA1yXta7vjhAVTHiQjYsNS\nOm6anYhI0izKeuQvejpfezWwunq8UNKdwF7Agk5pJUhGRE+VcZKNz9b2cuA220tGdkjaBnjI9pCk\nPYA9KWu6d9TkEKALgWuBvSUtkfSOpvKKiMEybHW1jadDnDmWdavaAIcBN0n6MXAx8B7bD42XR5O9\n28c1lXZEDK46ryTbxRnbJ46x7xLgkonmkep2RPSUEUNTP7CmawmSEdFz3VSl+0WCZET0lBFPeeZU\nF6NrCZIR0VNlMHmq2xERbfVgCFBtEiQjoqdsMeRcSU7KzBUr2fL8HzaW/oxNu1+jezKGVzZ/7/na\nu3/RbAYDNK3+WO557XMaz2PXGxvPYtobzpVkRMTYSsfN4ISewSlpREwL6biJiBjHUMZJRkSMLXfc\nRESMYzi92xERYysTXCRIRkSMyYg1uS0xImJsNgM1mLzJSXd3lvTfkhZLulXSyU3lFRGDRAx3ufWD\nJq8k1wIfsH2DpM2BhZKusv2TBvOMiD5nButKssmZyZdRVifD9mOSFgM7AgmSERu4dNyMUi0efgBw\nfS/yi4j+Zbpbv6ZfNB4kJW1GWVfilLHWt60WGz8JYC6bNF2ciJhiZUnZwekzbvSaV9JsSoC8wPbX\nxjrH9lm259ueP5uNmixORPQFMdTlNm5K0rmSlku6pWXfaZLuk7So2o5qOfYRSXdI+qmkV3VT2sbC\nuSQB5wCLbX+qqXwiYrCYWu+4+TzwOeALo/Z/2vYZrTsk7UNZanZfYAfgO5L2sj3UKYMmryQPAY4H\nXjpWRI+IDVddV5K2rwHGXTu78hrgy7ZX2/45cAdw0HgvarJ3+/vQJwOdIqJv2JrIleQ8SQtanp9l\n+6wuXvc+SW8FFlCGIj5MGV1zXcs5S6p9HQ1O62lETAul46br2xIftD1/glmcCXyiyuoTwCeBtzP2\nRdu4U/EnSEZEjzW7xo3t+5/OSTobuKx6ugTYueXUnYCl46U3OCM6I2JaKB036mqbDEnbtzx9HTDS\n8/0N4FhJG0naHdgTGHdRrVxJRkTP1XXHjaQLgcMpbZdLgFOBwyXtT4nHdwPvBrB9q6SLKHf9rQXe\nO17PNiRIRkSP1XnHje3jxth9Tofz/xr464nkkSAZET2XhcDWh4cbS7oX62IPuhmbb95o+sOPPdZo\n+nt8ZnGj6QOMWz+LjmxYM5wgGRExplLdTpCMiGirm7tp+kWCZET01MgQoEGRIBkRPZbqdkRER/2y\nfk03EiQjoqdK73aWlI2IGFOWb4iIGEeq24CkucA1wEZVPhfbPrWp/CJiMKR3+xmrgZfafrxa6+b7\nkr5l+7rxXhgR01t6twHbBh6vns6utnEnuIyI6c0WawcoSDa9WuJMSYuA5cBVtrPudkQ0Op9k3RoN\nkraHbO9PmQH4IEn7jT5H0kmSFkhasIbVTRYnIvpA05Pu1q0n17y2HwGuBo4Y41jW3Y7YwCRIApK2\nkbRl9Xhj4OXAbU3lFxGDYWSc5KAEySZ7t7cHzpM0kxKML7J92TiviYgNQMZJArZvAg5oKv2IGEw2\nrM2kuxER7fVLVbobCZIR0VODdu/24FzzRsS0YaurbTySzpW0XNItLfv+XtJtkm6SdGlLB/Jukp6Q\ntKja/qWbsiZIRkTPDaOuti58nmcPLbwK2M/2bwI/Az7ScuxO2/tX23u6ySBBMiJ6yq5vnKTta4CH\nRu270vba6ul1lJtZJi1BMiJ6TAwNz+hqA+aN3JFXbSdNMLO3A99qeb67pBslfU/Sod0kkI6biOi5\nbtobKw/anj+ZPCT9ObAWuKDatQzYxfYKSS8Cvi5pX9u/6pROXwVJP2cTVh86qc+jKxt980eNpd0r\nj77ptxpNf8uLb2w0fc2e02j6a/fZtdH0AfSDhxvPYzrrxXySkk4AjgZeVs1Ihu3VlCkcsb1Q0p3A\nXsCCTmn1VZCMiA2AS7tkUyQdAfwZ8Lu2V7Xs3wZ4yPaQpD2APYG7xksvQTIieq6u2xIlXQgcTmm7\nXAKcSunN3gi4ShLAdVVP9mHAxyWtBYaA99h+aMyEWyRIRkRPueq4qSUt+7gxdp/T5txLgEsmmkeC\nZET0XJPV7bolSEZEz02gd3vKJUhGRE/ZCZIRER0N0gQXCZIR0XNpk2xRzUy+ALjP9tFN5xcR/c2I\n4QGadLcXJT0ZWNyDfCJiQLjLrR80ve72TsDvAf/WZD4RMUBc33ySvdD0leRngA8Bw+1OWGfd7adW\nNlyciOgLA3Qp2eSSskcDy20v7HTeOutuz9m0qeJERB8ZpCvJJjtuDgGOkXQUMBfYQtL5tt/SYJ4R\n0ecMDA/3RwDsRmNXkrY/Ynsn27sBxwL/lQAZEaUqre62PpBxkhHRcxknOYrtq4Gre5FXRAyABMmI\niHb6p1OmGwmSEdF7uZKMiGjD4AHq3U6QjIgpkCAZEdFeqtsRER0kSE7OjMdXs8n/3tFY+kONpdw7\nW3395kbTn7HVlo2mP7Si2TWrZ9837uJ3621t4zlMcyODyWsg6VzK+trLbe9X7Xsu8BVgN+Bu4A22\nH1ZZOvGzwFHAKuBE2zeMl8fgTOoWEdOG3d3Whc8DR4za92Hgu7b3BL5bPQc4krLW9p7AScCZ3WTQ\ndZCUtFG350ZEdDSs7rZx2L4GGF19eA1wXvX4POC1Lfu/4OI6YEtJ24+Xx7hBUtJBkm4Gbq+ev1DS\nP45b+oiINuTuNmDeyFSK1XZSF8lvZ3sZQPVz22r/jsC9LectqfZ11E2b5D9Q6vxfrzL9saSXdPG6\niIhnm9hckQ/anl9TzmNdmo5bkm6q2zNs3zNq33ToA4mIKdHlDECT79y5f6QaXf1cXu1fAuzcct5O\nwNLxEusmSN4r6SDAkmZKOgX42cTKHBHRotmZyb8BnFA9PgH4j5b9b1VxMPDoSLW8k26q239IqXLv\nAtwPfKfaFxExOW0XdJkYSRcCh1PaLpcApwKnAxdJegfwC+D11emXU4b/3EEZAvS2bvIYN0jaXk6Z\nNDciYv3VOE7S9nFtDr1sjHMNvHeieYwbJCWdzRgXvrbH7WWSdDfwGKUNc22NDbARMcA0ze64+U7L\n47nA61i3G308L7H94IRKFRHT23QKkra/0vpc0heBqxorUUREH5nMbYm7A7t2ea6BKyUtbDcItHXd\n7af85CSKExGDZgKDyadcN22SD/PMxfEMyi1AH27/inUcYnuppG2BqyTdVt1G9DTbZwFnATxn1jZ9\n8rFERGNMV7cc9ouOQbKaNeOFwH3VruGqh6grtpdWP5dLuhQ4CLim86siYtoboMuhjtXtKiBeanuo\n2rp+a5I2lbT5yGPglcAt61XaiJgWplV1G/ihpAO7mXdtlO2AS8vFKLOAL9n+9kQLGBHTUJ8EwG60\nDZKSZtleC/wO8C5JdwIrKTeJ2/aBnRK2fRelqh4Rsa7pECSBHwIH8sxcbBER662fqtLd6BQkBWD7\nzh6VJSI2FNOkd3sbSe9vd9D2pxooT0RsAKbLleRMYDMGaYHciBgM0yRILrP98Z6VJCI2DNOtTTIi\nonbTJEg+az62xkkwZ3bPsx0kw6tWNZq+n1rTaPozdh133aX1YuW7fRCopkl3e6HtHTe2m1/lPSKi\nz3Vzx01ERL2mSXU7IqJ+06jjJiKiGQmSEREd1BQkJe0NtK6esAfwMWBL4F3AA9X+j9q+fDJ5JEhG\nRE+J+nq3bf8U2B9A0kzK3LeXUpaL/bTtM9Y3jwTJiOit5tokXwbcafse1TgUbDJr3ERErB93uU3M\nscCFLc/fJ+kmSedK2mqyRW00SEraUtLFkm6TtFjSbzeZX0QMiO6D5LyRhQKrrd2CgnOAY4CvVrvO\nBJ5HqYovAz452aI2Xd3+LPBt239QvYlNGs4vIgbABKrbD9qe38V5RwI32L4fYOQngKSzgcsmWsYR\njV1JStoCOAw4B8D2U7YfaSq/iBgg9Ve3j6Olqi1p+5Zjr2M91tdq8kpyD0r3+79LeiGwEDjZ9srW\nk6rL55MA5s7YrMHiRERfcL33bkvaBHgF8O6W3X8naf+SG3ePOjYhTbZJzqIs/3Cm7QMo6+M8a71u\n22fZnm97/pwZGzdYnIjoGzVeSdpeZXtr24+27Dve9m/Y/k3bx9heNtmiNhkklwBLbF9fPb+YEjQj\nYgM3SEvKNhYkbf8SuLcaEQ9lDNNPmsovIgZIM0OAGtF07/YfAxdUPdt3UUbBR8SGrI8CYDcaDZK2\nFwHddN9HxAZC9E9Vuhu5LTEiei5BMiKikwTJiIgOEiQjItroo+E93UiQjIjeS5CMiGhvkJaU7asg\n6bVrGXpgRXMZNL0ms5v/etSsZtcl95qnGk1/6I6fN5r+FUsXNZo+wKt22L/xPKa7VLcjItrJYPKI\niHEkSEZEjC133EREjEPDgxMlEyQjorfSJhkR0Vmq2xERnSRIRkS0N0hXkk2ulri3pEUt268kndJU\nfhExQDIzOdj+KWVhcCTNBO4DLm0qv4gYEDWvlti0XlW3XwbcafueHuUXEX2q7nGSku4GHgOGgLW2\n50t6LvAVYDfKkrJvsP3wZNJvcrXEVsfSsnB4K0knSVogacEaVveoOBExpezutu69xPb+tkeWi/kw\n8F3bewLfZYzlrLvVeJCsFgE7BvjqWMdb192ezUZNFyci+kAPlpR9DXBe9fg84LWTTagXV5JHAjfY\nvr8HeUVEv+u206YEyXkjNc1qO6lNildKWthyfDvbywCqn9tOtri9aJM8jjZV7YjYME2g4+bBlip0\nO4fYXippW+AqSbetV+FGafRKUtImwCuArzWZT0QMFg13t3XD9tLq53LKCJqDgPslbQ9Q/Vw+2bI2\nGiRtr7K9te1Hm8wnIgaIqa3jRtKmkjYfeQy8ErgF+AZwQnXaCcB/TLa4ueMmInquxiFA2wGXqqw6\nMAv4ku1vS/oRcJGkdwC/AF4/2QwSJCOi92oKkrbvAl44xv4VlPHZ6y1BMiJ6KpPuRkR0YmfS3YiI\njgYnRiZIRkTvpbodEdGOgVS318Pw0FSXoK957ZqpLkJfO2LXg3qQy1M9yGOaG5wY2YdBMiKmvVS3\nIyI6SO92REQ7fbQ0QzcSJCOip8pg8sGJkgmSEdF7WeMmIqK9XElGRLQzYG2STU+6+yeSbpV0i6QL\nJc1tMr+IGATl3u1utn7QWJCUtCPwf4H5tvcDZlJWTYyIDV39qyU2punq9ixgY0lrgE2ApQ3nFxH9\nzhNa42bKNXYlafs+4AzKrMDLgEdtX9lUfhExQAboSrLJ6vZWlLVvdwd2ADaV9JYxzjtpZLnINaxu\nqjgR0U+6X1J2yjXZcfNy4Oe2H7C9hrJi4otHn2T7LNvzbc+fzUYNFici+oWGh7va+kGTbZK/AA6u\nlpV9grLexIIG84uIQWAGajB5k22S1wMXAzcAN1d5ndVUfhExGISRu9vGTUvaWdJ/S1pcDTc8udp/\nmqT7JC2qtqMmW95Ge7dtnwqc2mQeETGA6uuUWQt8wPYN1frbCyVdVR37tO0z1jeD3HETEb1XU5C0\nvYwyegbbj0laDOxYS+KVRu+4iYh4lpE2yW42mDcy+qXaTmqXrKTdgAOA66td75N0k6Rzq9E2k5Ir\nyYjouQn0XD9oe/646UmbAZcAp9j+laQzgU9QQvIngE8Cb59MWRMkI6LH6h0oLmk2JUBeYPtrALbv\nbzl+NnDZZNNPdTsiesvUdseNJAHnAIttf6pl//Ytp70OuGWyxc2VZET0Xn3jJA8BjgdulrSo2vdR\n4DhJ+1NC8t3AuyebQYJkRPRcXZPu2v4+ZUWI0S6vJQP6LUhKaPacxpL3mqyXHNEX+mTyim70V5CM\niOnPhqHBuS8xQTIiei9XkhERHSRIRkS0YaBP1q/pRoJkRPSYwWmTjIgYm0nHTURERwPUJtn0utsn\nV2tu3yrplCbziogBkoXAQNJ+wLuAg4AXAkdL2rOp/CJiUHQZIKd7kAR+HbjO9irba4HvUW40j4gN\nmYHh4e62PtBkkLwFOEzS1tViYEcBOzeYX0QMigG6kmys48b2Ykl/C1wFPA78mLIexTqqmYZPApjL\nJk0VJyL6xmDdlthox43tc2wfaPsw4CHg9jHOeWbdbc1tsjgR0Q8M9nBXWz9odAiQpG1tL5e0C/D7\nwG83mV9EDIjccfO0SyRtDawB3mv74Ybzi4hB0Cftjd1oet3tQ5tMPyIGkN03PdfdyB03EdF7uZKM\niGjHeGhoqgvRtQTJiOitTJUWETGOPhne042sux0RPWXAw+5q64akIyT9VNIdkj5cd3kTJCOit1xN\nutvNNg5JM4F/Ao4E9qGst71PncVNdTsieq7GjpuDgDts3wUg6cvAa4Cf1JWB3Edd8ZIeAO6ZwEvm\nAQ82VJzfaoNiAAAC60lEQVSkP/3T70Ue/Zb+rra3aaow3ZD0bUq5uzEXeLLl+Vm2z2pJ6w+AI2y/\ns3p+PPBbtt9XV3n76kpyor88SQtsz2+qPEl/eqffizwGPf0m2D6ixuQ0VhY1pp82yYgYaEtYdwrG\nnYCldWaQIBkRg+xHwJ6Sdpc0BzgW+EadGfRVdXsSzhr/lKSf9Kc0j0FPv6/ZXivpfcAVwEzgXNu3\n1plHX3XcRDMkDQE3U74UFwMn2F41ybQOB/7U9tGSjgH2sX16m3O3BN5k+58nmMdpwOO2z5hMGSPq\nlOr2huEJ2/vb3g94CnhP60EVE/5bsP2NdgGysiXwRxNNN6KfJEhueP4HeL6k3SQtlvTPwA3AzpJe\nKelaSTdI+qqkzeDpOxpuk/R9yuTJVPtPlPS56vF2ki6V9ONqezFwOvA8SYsk/X113gcl/UjSTZL+\nsiWtP6/umvgOsHfPPo2IcSRIbkAkzaLcmXBztWtv4Au2DwBWAn8BvNz2gcAC4P2S5gJnA68GDgV+\nrU3y/wB8z/YLgQOBW4EPA3dWV7EflPRKYE/KAOD9gRdJOkzSiygN7gdQgvD/qfmtR0zaoHfcRHc2\nlrSoevw/wDnADsA9tq+r9h9Mua3rB5IA5gDXAi8Afm77dgBJ51Mt3DbKS4G3AtgeAh6VtNWoc15Z\nbTdWzzejBM3NgUtH2kkl1do7GbE+EiQ3DE/Y3r91RxUIV7buAq6yfdyo8/anvsG5Av6f7X8dlccp\nNeYRUatUt2PEdcAhkp4PIGkTSXsBtwG7S3pedd5xbV7/XeAPq9fOlLQF8BjlKnHEFcDbW9o6d5S0\nLXAN8DpJG0vanFK1j+gLCZIBgO0HgBOBCyXdRAmaL7D9JKV6/c2q46bdvfUnAy+RdDOwENjX9gpK\n9f0WSX9v+0rgS8C11XkXA5vbvgH4CrAIuITSJBDRFzJOMiKig1xJRkR0kCAZEdFBgmRERAcJkhER\nHSRIRkR0kCAZEdFBgmRERAf/H95G32jmoMjyAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f2c7eb9eb10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Let's try the Naive Bayes classifier as the SVM isn't working.\n",
    "#Naive Bayes is known to work suprisingly well for text data\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "\n",
    "import pylab as pl\n",
    "nb = MultinomialNB()\n",
    "nb.fit(train_feats_chi2, train_labels)\n",
    "preds = nb.predict(valid_feats_chi2)\n",
    "\n",
    "acc = accuracy(preds, np.array(valid_labels))\n",
    "print(\"Validation Accuracy: \" + str(acc))\n",
    "\n",
    "scores = cross_val_score(nb,train_feats_chi2, train_labels,  cv=5)\n",
    "print(\"Cross Validation Scores: \", scores)\n",
    "#Print confusion matrix\n",
    "#labels = [\"Class\" + str(i) for i in range(1,10)]\n",
    "nb_conf_mat = confusion_matrix(valid_labels, preds, labels = range(1,10))\n",
    "print(nb_conf_mat)\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "cax = ax.matshow(nb_conf_mat)\n",
    "pl.title('Confusion matrix of the classifier')\n",
    "fig.colorbar(cax)\n",
    "ax.set_xticklabels([''] + range(1,10))\n",
    "ax.set_yticklabels([''] + range(1,10))\n",
    "pl.xlabel('Predicted')\n",
    "pl.ylabel('True')\n",
    "pl.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation Accuracy: 59\n",
      "('Recall Score: ', 0.59699248120300752)\n",
      "('Precision Score: ', 0.62023083719576155)\n",
      "[[ 63   2   0  28   9   2  10   0   0]\n",
      " [  4  28   0   1   3   0  55   0   0]\n",
      " [  1   1   2   3   4   0   7   0   0]\n",
      " [ 30   1   0  83   2   2  19   0   0]\n",
      " [  8   1   0   9  11   2  17   0   0]\n",
      " [  8   0   0   9   1  24  13   0   0]\n",
      " [  2   3   0   3   0   0 183   0   0]\n",
      " [  0   0   0   1   0   0   3   0   0]\n",
      " [  0   1   0   3   0   0   0   0   3]]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nitin/anaconda2/envs/tensorflow/lib/python2.7/site-packages/sklearn/metrics/classification.py:1113: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.\n",
      "  'precision', 'predicted', average, warn_for)\n"
     ]
    }
   ],
   "source": [
    "from sklearn import linear_model\n",
    "from sklearn.metrics import recall_score, precision_score, f1_score\n",
    "log = linear_model.LogisticRegression()\n",
    "##One Hot encode training labels\n",
    "#train_labels = np.array(pd.get_dummies(train_df[\"Class\"]))\n",
    "#valid_labels = np.array(pd.get_dummies(valid_df[\"Class\"]))\n",
    "#train_labels = train_df[\"Class\"].tolist()\n",
    "#valid_labels = valid_df[\"Class\"].tolist()\n",
    "log.fit(train_feats, train_labels)\n",
    "scores = cross_val_score(log,train_feats, train_labels,  cv=5)\n",
    "preds = log.predict(valid_feats)\n",
    "acc = accuracy(preds, valid_labels)\n",
    "print(\"Validation Accuracy: \" + str(acc))\n",
    "\n",
    "#Print recall_score\n",
    "print(\"Recall Score: \", recall_score(valid_labels, preds, average = \"weighted\"))\n",
    "\n",
    "#Print precision score\n",
    "print(\"Precision Score: \", precision_score(valid_labels, preds,average = \"weighted\"))\n",
    "\n",
    "#Print confusion matrix\n",
    "log_conf_mat = confusion_matrix(valid_labels, preds, labels = range(1,10))\n",
    "print(log_conf_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "_cell_guid": "e5b1f826-c7c0-4805-b2fc-15a355b23b85",
    "_execution_state": "idle",
    "_uuid": "f7180db1958449e95f5340fb55ff57b5a7549a99"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation Accuracy: 55\n",
      "('Recall Score: ', 0.5518796992481203)\n",
      "('Precision Score: ', 0.54604401508279321)\n",
      "[[ 60   8   0  29  10   4   3   0   0]\n",
      " [  6  39   0   7   7   1  31   0   0]\n",
      " [  3   0   5   3   3   1   3   0   0]\n",
      " [ 27   8   3  80   2   5  10   0   2]\n",
      " [ 17   2   1   5  12   4   7   0   0]\n",
      " [  9   3   0   2   3  30   8   0   0]\n",
      " [  5  28   3   7   6   2 137   0   3]\n",
      " [  0   1   0   1   0   0   2   0   0]\n",
      " [  1   0   0   0   0   0   2   0   4]]\n"
     ]
    }
   ],
   "source": [
    "##Decision Tree Classifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "\n",
    "dtree = DecisionTreeClassifier(random_state=42).fit(train_feats,train_labels)\n",
    "\n",
    "preds = dtree.predict(valid_feats)\n",
    "\n",
    "#scores = cross_val_score(dtree, train_feats, train_df[\"Class\"], cv=5)\n",
    "#print(scores)\n",
    "\n",
    "acc = accuracy(preds, valid_labels)\n",
    "print(\"Validation Accuracy: \" + str(acc))\n",
    "\n",
    "#Print recall_score\n",
    "print(\"Recall Score: \", recall_score(valid_labels, preds, average = \"weighted\"))\n",
    "\n",
    "#Print precision score\n",
    "print(\"Precision Score: \", precision_score(valid_labels, preds,average = \"weighted\"))\n",
    "\n",
    "#Print confusion matrix\n",
    "dtree_conf_mat = confusion_matrix(valid_labels, preds, labels = range(1,10))\n",
    "print(dtree_conf_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Iter       Train Loss   Remaining Time \n",
      "         1        4974.5918            3.63m\n",
      "         2        4536.5290            3.67m\n",
      "         3        4206.6064            3.65m\n",
      "         4        3942.7054            3.65m\n",
      "         5        3729.3517            3.66m\n",
      "         6        3549.2955            3.65m\n",
      "         7        3381.4771            3.60m\n",
      "         8        3236.0427            3.59m\n",
      "         9        3111.9499            3.54m\n",
      "        10        2998.7230            3.51m\n",
      "        11        2897.0265            3.49m\n",
      "        12        2801.1299            3.49m\n",
      "        13        2723.3643            3.44m\n",
      "        14        2654.1016            3.41m\n",
      "        15        2579.3946            3.39m\n",
      "        16        2508.6383            3.37m\n",
      "        17        2450.0542            3.35m\n",
      "        18        2390.2860            3.32m\n",
      "        19        2344.2317            3.29m\n",
      "        20        2293.3216            3.26m\n",
      "        21        2244.0903            3.24m\n",
      "        22        2204.7331            3.21m\n",
      "        23        2165.8532            3.18m\n",
      "        24        2133.1761            3.14m\n",
      "        25        2093.2392            3.11m\n",
      "        26        2061.6675            3.09m\n",
      "        27        2032.0267            3.07m\n",
      "        28        2000.9099            3.07m\n",
      "        29        1978.3453            3.03m\n",
      "        30        1954.6575            3.00m\n",
      "        31        1930.5618            2.97m\n",
      "        32        1913.4937            2.94m\n",
      "        33        1891.7920            2.91m\n",
      "        34        1868.4688            2.89m\n",
      "        35        1850.4214            2.87m\n",
      "        36        1831.1416            2.85m\n",
      "        37        1816.6722            2.82m\n",
      "        38        1792.9645            2.80m\n",
      "        39        1776.8920            2.77m\n",
      "        40        1759.8809            2.74m\n",
      "        41        1738.6808            2.72m\n",
      "        42        1725.1971            2.69m\n",
      "        43        1707.4410            2.67m\n",
      "        44        1691.8130            2.65m\n",
      "        45        1676.8917            2.63m\n",
      "        46        1662.7703            2.61m\n",
      "        47        1648.9556            2.58m\n",
      "        48        1639.5186            2.55m\n",
      "        49        1625.4704            2.53m\n",
      "        50        1613.0117            2.51m\n",
      "        51        1602.4717            2.48m\n",
      "        52        1586.7223            2.46m\n",
      "        53        1576.7256            2.43m\n",
      "        54        1564.7832            2.41m\n",
      "        55        1553.8730            2.39m\n",
      "        56        1539.7800            2.37m\n",
      "        57        1528.9324            2.36m\n",
      "        58        1517.9676            2.34m\n",
      "        59        1505.5179            2.33m\n",
      "        60        1496.1611            2.30m\n",
      "        61        1488.7444            2.28m\n",
      "        62        1481.8571            2.26m\n",
      "        63        1470.0748            2.24m\n",
      "        64        1457.3893            2.22m\n",
      "        65        1447.6901            2.20m\n",
      "        66        1439.5917            2.19m\n",
      "        67        1430.8691            2.17m\n",
      "        68        1422.4787            2.15m\n",
      "        69        1415.0161            2.13m\n",
      "        70        1406.2744            2.11m\n",
      "        71        1398.7799            2.09m\n",
      "        72        1392.2461            2.07m\n",
      "        73        1386.1360            2.05m\n",
      "        74        1378.3144            2.03m\n",
      "        75        1371.1193            2.02m\n",
      "        76        1365.3095            2.00m\n",
      "        77        1356.4876            1.98m\n",
      "        78        1349.1261            1.96m\n",
      "        79        1341.8014            1.95m\n",
      "        80        1333.4437            1.93m\n",
      "        81        1326.6398            1.91m\n",
      "        82        1319.0466            1.89m\n",
      "        83        1310.2289            1.87m\n",
      "        84        1300.7138            1.86m\n",
      "        85        1294.0138            1.84m\n",
      "        86        1287.2905            1.83m\n",
      "        87        1282.1147            1.81m\n",
      "        88        1276.2142            1.79m\n",
      "        89        1268.8413            1.77m\n",
      "        90        1261.8974            1.75m\n",
      "        91        1253.2778            1.74m\n",
      "        92        1247.9731            1.72m\n",
      "        93        1241.7013            1.71m\n",
      "        94        1236.1106            1.69m\n",
      "        95        1230.1337            1.67m\n",
      "        96        1223.6349            1.66m\n",
      "        97        1218.7014            1.64m\n",
      "        98        1213.6329            1.62m\n",
      "        99        1207.5333            1.60m\n",
      "       100        1201.1577            1.59m\n",
      "       101        1194.7371            1.57m\n",
      "       102        1190.8888            1.55m\n",
      "       103        1186.9822            1.54m\n",
      "       104        1181.9202            1.52m\n",
      "       105        1177.4906            1.50m\n",
      "       106        1171.3762            1.48m\n",
      "       107        1166.8975            1.47m\n",
      "       108        1163.3649            1.45m\n",
      "       109        1158.0117            1.43m\n",
      "       110        1152.4247            1.42m\n",
      "       111        1146.6470            1.40m\n",
      "       112        1139.0183            1.39m\n",
      "       113        1134.1732            1.37m\n",
      "       114        1130.7080            1.35m\n",
      "       115        1126.3119            1.34m\n",
      "       116        1120.4747            1.32m\n",
      "       117        1115.1224            1.30m\n",
      "       118        1110.7094            1.29m\n",
      "       119        1107.3927            1.27m\n",
      "       120        1103.6686            1.26m\n",
      "       121        1098.6572            1.24m\n",
      "       122        1094.5312            1.22m\n",
      "       123        1090.9974            1.21m\n",
      "       124        1086.8761            1.19m\n",
      "       125        1083.1186            1.17m\n",
      "       126        1077.3912            1.16m\n",
      "       127        1072.8126            1.14m\n",
      "       128        1069.4090            1.13m\n",
      "       129        1064.9256            1.11m\n",
      "       130        1061.3453            1.09m\n",
      "       131        1057.7446            1.08m\n",
      "       132        1054.6418            1.06m\n",
      "       133        1050.4524            1.05m\n",
      "       134        1046.1680            1.03m\n",
      "       135        1042.3626            1.01m\n",
      "       136        1039.7612           59.84s\n",
      "       137        1036.3133           58.91s\n",
      "       138        1031.2908           58.00s\n",
      "       139        1027.6595           57.07s\n",
      "       140        1023.4848           56.12s\n",
      "       141        1021.0336           55.15s\n",
      "       142        1017.6279           54.18s\n",
      "       143        1013.3226           53.23s\n",
      "       144        1009.8281           52.27s\n",
      "       145        1006.7051           51.35s\n",
      "       146        1003.5722           50.42s\n",
      "       147         999.4139           49.50s\n",
      "       148         996.2091           48.56s\n",
      "       149         992.7342           47.62s\n",
      "       150         990.1997           46.65s\n",
      "       151         987.3440           45.68s\n",
      "       152         984.0616           44.72s\n",
      "       153         981.1640           43.77s\n",
      "       154         978.5397           42.82s\n",
      "       155         974.4711           41.89s\n",
      "       156         971.3298           40.96s\n",
      "       157         969.1005           40.02s\n",
      "       158         965.2337           39.09s\n",
      "       159         962.1873           38.13s\n",
      "       160         959.7830           37.17s\n",
      "       161         956.7048           36.22s\n",
      "       162         953.6558           35.27s\n",
      "       163         951.5039           34.33s\n",
      "       164         947.9702           33.41s\n",
      "       165         944.2674           32.50s\n",
      "       166         941.3818           31.56s\n",
      "       167         937.5965           30.72s\n",
      "       168         934.5747           29.90s\n",
      "       169         931.7481           29.08s\n",
      "       170         929.0662           28.18s\n",
      "       171         926.0990           27.28s\n",
      "       172         922.8545           26.38s\n",
      "       173         920.5877           25.46s\n",
      "       174         918.1095           24.56s\n",
      "       175         915.6882           23.64s\n",
      "       176         911.7047           22.73s\n",
      "       177         908.6330           21.81s\n",
      "       178         905.8664           20.85s\n",
      "       179         904.1129           19.89s\n",
      "       180         901.7283           18.95s\n",
      "       181         898.3741           18.00s\n",
      "       182         896.5678           17.05s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       183         894.2296           16.10s\n",
      "       184         891.6248           15.16s\n",
      "       185         889.6805           14.20s\n",
      "       186         887.1495           13.24s\n",
      "       187         884.8553           12.29s\n",
      "       188         882.7021           11.34s\n",
      "       189         880.1496           10.39s\n",
      "       190         878.2472            9.45s\n",
      "       191         875.5859            8.50s\n",
      "       192         872.5928            7.56s\n",
      "       193         869.5961            6.61s\n",
      "       194         867.1853            5.67s\n",
      "       195         864.9247            4.72s\n",
      "       196         863.1722            3.77s\n",
      "       197         861.5729            2.83s\n",
      "       198         859.2164            1.89s\n",
      "       199         856.8074            0.94s\n",
      "       200         854.0381            0.00s\n",
      "Validation Accuracy: 62\n",
      "('Recall Score: ', 0.62857142857142856)\n",
      "('Precision Score: ', 0.62815970788053466)\n",
      "['class1', 'class2', 'class3', 'class4', 'class5', 'class6', 'class7', 'class8', 'class9']\n",
      "   ID  class1  class2  class3  class4  class5  class6  class7  class8  class9\n",
      "0   0       0       0       0       0       0       0       1       0       0\n",
      "1   1       0       0       0       1       0       0       0       0       0\n",
      "2   2       0       0       0       0       0       0       1       0       0\n",
      "3   3       0       0       0       0       0       0       1       0       0\n",
      "4   4       0       1       0       0       0       0       0       0       0\n",
      "Index([u'ID', u'class1', u'class2', u'class3', u'class4', u'class5', u'class6',\n",
      "       u'class7', u'class8', u'class9'],\n",
      "      dtype='object')\n",
      "Submission file created.\n"
     ]
    }
   ],
   "source": [
    "##Trying GBM before heading to XGBoost\n",
    "\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "\n",
    "\n",
    "gbm = GradientBoostingClassifier(verbose = 1, max_depth=3,n_estimators=200).fit(train_feats_chi2,train_labels)\n",
    "preds = gbm.predict(valid_feats_chi2)\n",
    "test_preds = gbm.predict(test_feats_chi2)\n",
    "#scores = cross_val_score(dtree, train_feats, train_df[\"Class\"], cv=5)\n",
    "#print(scores)\n",
    "\n",
    "acc = accuracy(preds, valid_labels)\n",
    "print(\"Validation Accuracy: \" + str(acc))\n",
    "\n",
    "#Print recall_score\n",
    "print(\"Recall Score: \", recall_score(valid_labels, preds, average = \"weighted\"))\n",
    "\n",
    "#Print precision score\n",
    "print(\"Precision Score: \", precision_score(valid_labels, preds,average = \"weighted\"))\n",
    "\n",
    "make_submission(predictions=test_preds,test_df=test_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([   1,    3,    4, ..., 7996, 8000, 8001])"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Parameter values for parameter (scale_pos_weight) need to be a sequence.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-28-df5d71129f24>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     42\u001b[0m }\n\u001b[1;32m     43\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 44\u001b[0;31m \u001b[0mgrid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGridSearchCV\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparam_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mfit_params\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfit_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     45\u001b[0m \u001b[0mgrid\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_feats_chi2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtrain_labels_xgb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nitin/anaconda2/envs/tensorflow/lib/python2.7/site-packages/sklearn/grid_search.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, estimator, param_grid, scoring, fit_params, n_jobs, iid, refit, cv, verbose, pre_dispatch, error_score)\u001b[0m\n\u001b[1;32m    810\u001b[0m             refit, cv, verbose, pre_dispatch, error_score)\n\u001b[1;32m    811\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparam_grid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparam_grid\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 812\u001b[0;31m         \u001b[0m_check_param_grid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparam_grid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    813\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    814\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nitin/anaconda2/envs/tensorflow/lib/python2.7/site-packages/sklearn/grid_search.pyc\u001b[0m in \u001b[0;36m_check_param_grid\u001b[0;34m(param_grid)\u001b[0m\n\u001b[1;32m    346\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mTrue\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcheck\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    347\u001b[0m                 raise ValueError(\"Parameter values for parameter ({0}) need \"\n\u001b[0;32m--> 348\u001b[0;31m                                  \"to be a sequence.\".format(name))\n\u001b[0m\u001b[1;32m    349\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    350\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Parameter values for parameter (scale_pos_weight) need to be a sequence."
     ]
    }
   ],
   "source": [
    "\n",
    "import xgboost as xgb\n",
    "from xgboost.sklearn import XGBClassifier\n",
    "from sklearn import cross_validation, metrics   #Additional scklearn functions\n",
    "from sklearn.grid_search import GridSearchCV   #Perforing grid search\n",
    "\n",
    "import matplotlib.pylab as plt\n",
    "%matplotlib inline\n",
    "from sklearn.metrics import accuracy_score\n",
    "from matplotlib.pylab import rcParams\n",
    "rcParams['figure.figsize'] = 12, 4\n",
    "\n",
    "train_labels_onehot = pd.get_dummies(train_labels)\n",
    "valid_labels_onehot = pd.get_dummies(valid_labels)\n",
    "\n",
    "train_labels_xgb = [i-1 for i in train_labels]\n",
    "valid_labels_xgb = [i-1 for i in valid_labels]\n",
    "\n",
    "param_dict = {\n",
    "    \"learning_rate\" :0.1,\n",
    "    \"n_estimators\":1000,\n",
    "    \"max_depth\":5,\n",
    "    \"min_child_weight\":1,\n",
    "    \"gamma\":0,\n",
    "    \"subsample\":0.8,\n",
    "    \"colsample_bytree\":0.8,\n",
    "    \"objective\": 'multi:softmax',\n",
    "    \"nthread\":4,\n",
    "    \"scale_pos_weight\":1,\n",
    "    \"seed\":42\n",
    "}\n",
    "\n",
    "alg = XGBClassifier()\n",
    "alg.set_params(**param_dict)\n",
    "\n",
    "\n",
    "\n",
    "fit_params = {\n",
    "'early_stopping_rounds': 30,\n",
    "'eval_metric': 'mae',\n",
    "'eval_set': [[train_feats_chi2,train_labels]]\n",
    "}\n",
    "\n",
    "grid = GridSearchCV(alg, param_dict, cv=5,fit_params=fit_params)\n",
    "grid.fit(train_feats_chi2,train_labels_xgb)\n",
    "\n",
    "y_pred = model.predict(valid_feats_chi2)\n",
    "predictions = [round(value) for value in y_pred]\n",
    "# evaluate predictions\n",
    "accuracy = accuracy_score(valid_labels_xgb, predictions)\n",
    "print(\"Accuracy: %.2f%%\" % (accuracy * 100.0))\n",
    "\n",
    "\n",
    "\n",
    "#predictors = [feature_names[i] for i in ch2.get_support(indices=True)]\n",
    "#print(predictors[0:10])\n",
    "\"\"\"\n",
    "useTrainCV = True\n",
    "cv_folds = 5\n",
    "early_stopping_rounds = 50\n",
    "\n",
    "    \n",
    "if useTrainCV:\n",
    "    xgb_param = alg.get_xgb_params()\n",
    "    xgtrain = xgb.DMatrix(train_feats, label=train_labels)\n",
    "    cvresult = xgb.cv(xgb_param, xgtrain, num_boost_round=alg.get_params()['n_estimators'], nfold=cv_folds,\n",
    "        metrics='mlogloss', early_stopping_rounds=early_stopping_rounds,seed = 42)\n",
    "    alg.set_params(n_estimators=cvresult.shape[0])\n",
    "\n",
    "        \n",
    "#Predict training set:\n",
    "dtrain_predictions = alg.predict(valid_feats)\n",
    "dtrain_predprob = alg.predict_proba(valid_feats)[:,1]\n",
    "        \n",
    "#Print model report:\n",
    "print(\"\\nModel Report\")\n",
    "print(\"Accuracy : %.4g\" % metrics.accuracy_score(valid_labels, dtrain_predictions))\n",
    "print(\"AUC Score (Train): %f\" % metrics.roc_auc_score(valid_labels, dtrain_predprob))\n",
    "                    \n",
    "feat_imp = pd.Series(alg.booster().get_fscore()).sort_values(ascending=False)\n",
    "feat_imp.plot(kind='bar', title='Feature Importances')\n",
    "plt.ylabel('Feature Importance Score')\n",
    "#Fit the algorithm on the data\n",
    "alg.fit(train_feats,train_labels_xgb, eval_metric='mlogloss',verbose = True)\n",
    "\n",
    "\n",
    "    \n",
    "model = XGBClassifier()\n",
    "model.fit(train_feats, train_labels_xgb, verbose = True)\n",
    "# make predictions for test data\n",
    "y_pred = model.predict(valid_feats)\n",
    "predictions = [round(value) for value in y_pred]\n",
    "# evaluate predictions\n",
    "accuracy = accuracy_score(valid_labels_xgb, predictions)\n",
    "print(\"Accuracy: %.2f%%\" % (accuracy * 100.0))\n",
    "\"\"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 1 candidates, totalling 5 fits\n",
      "[CV] n_estimators=500, learning_rate=0.2, max_depth=5 ................\n",
      "[CV]  n_estimators=500, learning_rate=0.2, max_depth=5, score=0.687266 - 4.5min\n",
      "[CV] n_estimators=500, learning_rate=0.2, max_depth=5 ................\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  4.5min remaining:    0.0s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CV]  n_estimators=500, learning_rate=0.2, max_depth=5, score=0.680451 - 4.9min\n",
      "[CV] n_estimators=500, learning_rate=0.2, max_depth=5 ................\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  9.4min remaining:    0.0s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CV]  n_estimators=500, learning_rate=0.2, max_depth=5, score=0.636535 - 4.7min\n",
      "[CV] n_estimators=500, learning_rate=0.2, max_depth=5 ................\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed: 14.1min remaining:    0.0s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CV]  n_estimators=500, learning_rate=0.2, max_depth=5, score=0.623352 - 4.7min\n",
      "[CV] n_estimators=500, learning_rate=0.2, max_depth=5 ................\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed: 18.9min remaining:    0.0s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CV]  n_estimators=500, learning_rate=0.2, max_depth=5, score=0.696970 - 5.5min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed: 24.4min remaining:    0.0s\n",
      "[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed: 24.4min finished\n"
     ]
    },
    {
     "ename": "XGBoostError",
     "evalue": "need to call fit beforehand",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mXGBoostError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-40-f4489e0ec932>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_feats_chi2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrain_labels_xgb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m \u001b[0mfeat_imp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mxgb_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbooster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_fscore\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0mfeat_imp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkind\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'bar'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtitle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'Feature Importances'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mylabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Feature Importance Score'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/nitin/anaconda2/envs/tensorflow/lib/python2.7/site-packages/xgboost/sklearn.pyc\u001b[0m in \u001b[0;36mbooster\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    161\u001b[0m         \"\"\"\n\u001b[1;32m    162\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_Booster\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 163\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mXGBoostError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'need to call fit beforehand'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    164\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_Booster\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mXGBoostError\u001b[0m: need to call fit beforehand"
     ]
    }
   ],
   "source": [
    "import xgboost\n",
    "from sklearn.metrics import make_scorer, accuracy_score\n",
    "from sklearn.grid_search import GridSearchCV\n",
    "\n",
    "# Load and process train users\n",
    "\n",
    "xgb_model = xgboost.XGBClassifier(objective=\"multi:softmax\", nthread=-1,subsample=0.8)\n",
    "\n",
    "clf = GridSearchCV(\n",
    "    xgb_model,\n",
    "    {\n",
    "        'max_depth': [5],\n",
    "        'n_estimators': [500],\n",
    "        'learning_rate': [0.2]\n",
    "    },\n",
    "    cv=5,\n",
    "    verbose=6,\n",
    "    n_jobs=1,\n",
    "    scoring=make_scorer(accuracy_score)\n",
    ")\n",
    "\n",
    "\n",
    "clf.fit(train_feats_chi2, train_labels_xgb)\n",
    "feat_imp = pd.Series(xgb_model.booster().get_fscore()).sort_values(ascending=False)\n",
    "feat_imp.plot(kind='bar', title='Feature Importances')\n",
    "plt.ylabel('Feature Importance Score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'abstract germline mutations that inactivate the tumor suppressor gene brca are associated with an increased risk of cancers of the breast and other tissues but the functional consequence of many missense variants found in the human population is uncertain several predictive methods have been proposed to distinguish cancer predisposing missense mutations from harmless polymorphisms including a small colony phenotype scp assay performed in the model organism yeast saccharomyces cerevisiae the goal of this study was to further evaluate this colony size assay we constructed missense mutations throughout the c terminal amino acid residues of brca mutated proteins were expressed in yeast and evaluated using the scp assay we conclude there is as yet no evidence the assay can identify inactivating mutations upstream of the brct repeats however within and between the brct repeats results of the assay are in general agreement with predictions based on structural modeling other in vitro and in vivo assays and cross species sequence conservation thus the yeast assay appears to provide confirmatory in vivo evidence to aid in characterizing some brca missense variants introduction approximately of breast cancer cases can be attributed to germline mutations in the brca gene which also confer increased risk of ovarian and prostate cancer brca has the hallmarks of a tumor suppressor gene the wild type allele is characteristically lost in tumors of germline mutation carriers also transfection of wild type brca into human tumor cell lines results in growth inhibition through cell cycle arrest thus germline brca mutations are expected to result from gene inactivation indeed over of brca sequence variants catalogued in the breast cancer information core bic database are frameshift or nonsense mutations resulting in premature translation termination regulatory mutations or splice site alterations since the deletion of only eleven c terminal amino acid residues is known to result in predisposition to breast cancer most or all of these mutations very likely inactivate brca function as well as has been confirmed in numerous cases the risk associated with missense variants of which over have been described is much more difficult to predict many at risk individuals currently receive a brca genetic test result that is ambiguous in a few cases the testing of samples from many family members provides evidence of linkage between a particular missense allele and increased cancer risk however genetic evidence is often either unavailable or inconclusive about half of the mutations reported to the bic are from a single family or individual and because noninherited breast cancer is common clusters of affected family members may occur by chance this uncertainty has substantial health implications confirmed brca mutation carriers may elect presymptomatic interventions such as prophylactic mastectomy or oophorectomy and or more frequent and aggressive screening should cancer be discovered increasing evidence suggests that brca associated tumors are genetically and phenotypically distinct from sporadic tumors and respond differently to treatment thus a test or series of tests capable of distinguishing cancer predisposing brca alleles from neutral polymorphisms would be of great value despite extensive study the precise biochemical function of the brca protein is unclear through its association with proteins involved in dna repair transcription and chromatin modification it has been proposed to play a role in the maintenance of genome integrity this role is likely to be multi faceted making a biochemical assay for brca functionality challenging in the absence of conclusive genetic or biochemical means of classification most missense variants of brca in the bic database are listed as unclassified variants tentative predictions of functionality can be made on the basis of sequence conservation between www landesbioscience com cancer biology therapy species and structural modeling complete sequences of brca orthologs are available for the chimpanzee dog mouse rat cow chicken and african clawed frog xenopus laevis alteration of residues highly conserved between these species and humans are expected to be more likely to affect brca function a comparison of several additional species has been made for an internal region of the gene structural predictions can be made for two regions of the brca protein the n terminal ring finger and the two c terminal brct domains the three dimensional structures of the two brct domains of the human and rat have been solved by x ray crystallography each brct domain contains a four stranded parallel \\u03b2 sheet surrounded by three \\u03b1 helices sequence alterations predicted to disrupt the folding of one of the brct domains can be considered more likely to affect the protein s function and therefore to be diseaseassociated direct in vitro biophysical measurement of the structural consequences of some missense mutations has also been reported while each of these predictive methods is valuable an in vivo assay would provide additional information it has been shown that transfected wild type brca decreases sensitivity to ionizing radiation of a brca mouse cell line whereas disease associated missense alleles do not missense alleles have also been shown to affect chromatin unfolding in an engineered mammalian cell line these assays shows great promise but are complex and results from them have as yet been reported only for a small number of already classified alleles other simpler assays may also contribute to characterizing the growing number of sequence variants two in vivo activities associated with the c terminal region of brca have been identified when this region is expressed in the yeast saccharomyces cerevisiae first fused to a dna binding domain it allows transcriptional activation of heterologous promoters second it confers a small colony phenotype scp on yeast transformants this growth phenotype discovered serendipitously in the course of two hybrid experiments is enhanced by fusion of the brct domains to the transcription activation domain of yeast gal and the sv large t antigen nuclear localization signal expression in yeast of human bard which contains two brct domains also results in growth inhibition a further upstream region of brca amino acid residues also exerts a negative effect on yeast growth as well as a transcription activation function in yeast and mammalian cells thus these two yeast phenotypes may be mechanistically related both of the brca yeast phenotypes are abolished by truncations or missense mutations known to be disease associated but are insensitive to mutations known to represent benign polymorphisms the phenotypic differences do not result from altered protein stability as all the variants were shown by western blotting to be present in equivalent amounts although the biochemical basis of the small colony phenotype is not yet understood its manifestation appears to rely on the expression of a functionally intact brca c terminal region apparently both brct domains are required since truncation of the c terminal domain or missense mutations in either abrogates the phenotype it has therefore been proposed that the small colony and transcriptional activation phenotypes may offer simple assays predicting the functional status of brca sequence variants in the brct region the purposes of the experiments reported here were to further characterize the scp assay and to evaluate its potential in brca sequence variant classification materials and methods site directed mutagenesis to construct each desired mutation two complementary oligonucleotide primers were synthesized containing the required base pair change s in the center one of two alternate methods was then used the splice overlap extension method was performed as described by vallejo et al using the mutagenic primers and the flanking primers actseq ctagcttgggtggtcatatg and actseq cggggtttttcagtatctacg and the plasmid padbrca as template the mutated brca pcr product was digested with ecori and bamhi and inserted into the same sites of the plasmid pact alternatively the quick change method was applied directly on the plasmid pad brca as recommended by the manufactuer stratagene inc dna sequencing for each plasmid the entire brca insert was sequenced on both strands to confirm the presence of the targeted nucleotide substitution and the absence of any untargeted mutations using either an abi or abi automated dna sequencer applied biosystems inc sequences were assembled using sequencher software gene codes corp yeast transformation the s cerevisiae strain l mata his d trp leu ade lys lexaop his ura lexaop lacz was transformed using standard lithium acetate procedures and spread on synthetic yeast medium lacking leucine colonies were grown approximately hours at c before restreaking colony size quantification two randomly selected colonies from each initial transformation plate were restreaked and incubated at c exactly hours two or more colonies from each streak were separately resuspended in sterile distilled water using a pipettor the cells were counted using a hemocytometer results using pcr mediated site directed mutagenesis we introduced missense variants within the c terminal amino acid residues of brca table each variant was inserted into a translational fusion expression vector as previously described each plasmid as well as several controls was transformed into the yeast strain l expression of the scp was found to be quite strain dependent and l provided a clearly visible distinction between small and large colonies fig colony size was determined as described in materials and methods the results are presented in figure mutations upstream of the brct repeats humphrey et al reported that most of the growth inhibition is associated with expression of the c terminal amino acid residues which contain the brct domains however expression of a longer brca fragment c terminal residues caused an additional two fold reduction in colony size mammalian cross species sequence conservation in this residue region upstream of brct n the n terminal brct is considerably lower than within the brct domains and no meaningful sequence alignment is apparent between human brca and the chicken and xenopus orthologs no confirmed disease associated missense allele has been mapped to this region we tested whether missense mutations within this region could detectably alter the scp one previously tested mutation p l and ten new mutations in this region were tested as shown in table seven of these eleven alleles have been detected in humans the other four including the double mutation lk ee were selected to drastically alter from hydrophobic or basic to acidic residues conserved between the six sequenced mammalian orthologs as shown in figure none of these eleven mutations caused a significant increase in the size of the yeast colonies one of the mutations s c has been reported to abolish transcriptional activation in yeast we conclude from our results that there is as yet no evidence that the scp assay can identify inactivating missense mutations in the region upstream of the brct domains in contrast at least one predicted inactivating mutation has been mapped to this region using the yeast transcription activation assay yeast assay for brca missense variant characterization cancer biology therapy vol issue www landesbioscience com cancer biology therapy yeast assay for brca missense variant characterization mutations within the brct region eleven of the eighteen novel point mutations within the brct region showed increased yeast colony size compared with yeast expressing wild type brca indicating a probable disruption of brct structure and or function among these mutations there was an approximately fold range of colony size the previouslytested control residue truncation allele y x showed the greatest difference from wild type brca it is possible that colony size correlates with severity of the missense mutations but further study will be required to determine this the correlation between level of cross species sequence conservation and effect of mutation on the scp is striking mutations altering eight of the nine residues conserved in all eight species table result in abrogation of the scp v m is the one exception conversely mutations altering only two and of the nine positions at which multiple residues are allowed result in such abrogation substitutions at position can result in opposite effects in the yeast transcription activation assay depending on the substituted residue m k abolishes transcription activation whereas m i has no effect m i also has no effect on the scp whereas our results suggest m k and m t are deleterious thus it appears mutations at this position need to be especially carefully evaluated furthermore results of the scp assay for these eighteen alleles are consistent with nearly every other prediction that has been made based on structural modeling in vitro biophysical characterization and or transcription activation phenotype see allele specific references in table the only exceptions are four disagreements with the predictions of mirkovic et al which are based on a set of structure based rules d e m t and v m were predicted by these rules to be likely benign replacements whereas the scp assay indicates possible functional and or structural defects conversely the structure based algorithm predicts a deleterious effect of v d a mutation that shows a similar scp to wild type brca of particular interest is the modest aspartate to glutamate substitution at residue between the two brct repeats although the human crystal structure in the region of this residue is unresolved the corresponding region of the rat structure shows a short \\u03b2 sheet predicted to stabilize the interface of the two brct repeats determination of the structural consequences of this modest substitution may reveal the basis for its effect on the scp and resolve the status of its likely disease association discussion our results suggest that the structural integrity of the brct region approximately the most c terminal amino acid residues of brca is the most critical determinant in the yeast scp assay none of the tested missense mutations in the residue region upstream of brct c including ones affecting conserved positions and one that abolishes transcription activation in yeast and mammalian cells has a detectable effect on growth thus there is as yet no evidence that the scp assay may be useful in characterizing variants upstream of residue within the brct region mutations that abolish the scp are interspersed with those having no effect seven mutations at five positions display an scp similar to wild type brca all but one v m alter residues incompletely conserved between the orthologs of eight vertebrate species and in fact three result in substitution of a residue allowed in another species at the corresponding position t i v a p a with one exception v d the scp assay predictions for these alleles are consistent with all other available predictions thus the evidence favors assignment of these variants as benign polymorphisms however these amino acid substitutions may affect functional attributes not detectable by the scp assay therefore figure relative colony sizes the yeast strain l was transformed with various plasmids individual transformants were restreaked into sectors of the same plate and incubated at c for hours clockwise starting at top pad brca wild type g v k e pact vector figure quantification of colony sizes number of cells per yeast colony transformed with the vector pact pad brca wild type brca or the brca plasmid carrying the indicated mutation error bars indicate standard deviation its value in characterizing polymorphisms is uncertain elucidation of the biochemical basis for the scp and its relation to brca s function in human cells may help clarify this issue in contrast we identified eleven variants at nine positions within brct n brct c or in the interval between them that alleviate the scp all eleven affect residues completely conserved between the brca genes of all eight vertebrates with two exceptions m and m are both substituted with i in the chicken all have been predicted by various means to disrupt folding and or function of the brct domain thus the scp assay provides confirmatory in vivo evidence for the deleterious nature of these mutations although caution must be exercised in applying these results directly to clinical decision making we conclude that the yeast scp assay may be useful in conjunction with other information to evaluate selected alleles this confirmation of the potential value of the scp assay highlights the need to understand its biochemical basis this understanding table mutations characterized in this study secondary structure predicted structural references for functional mutation allowed residues detected in humans consequences predictions upstream of brct l e l n unknown k e k n lk ee lk n k e q k r y a v a e y s c s n y a d a n m v m s v r y m t m s v r y p l p y a it a s y brct region m t m i f y adjacent to \\u03b2 sheet brct n m k m i f n v m v y s f s y t a t v i y \\u03b1 helix brct n t i t v i n v a v y between brct domains d e d y d g d y m r m y \\u03b1 helix brct c surface disrupts interface m t m i y \\u03b1 helix brct c intrahelical destabilizes fold g v g y loop between \\u03b1 helix and \\u03b2 sheet brct c incorrect folding v a v t a s n loop between \\u03b2 sheets and brct c v d v t a s y p a p l a s y v a v y \\u03b2 sheet brct c destabilizes fold v g v y n y n s a t r n \\u03b1 helix brct c v m v y \\u03b2 sheet brct c destabilizes fold y x y y residue truncation those italicized and underlined abrogate the small colony phenotype and thus are predicted to negatively affect brca structure and or function upstream of brct allowed residues in mammals within brct region allowed residues in mammals chicken and xenopus this reference provides information on a different substitution at this residue or an adjacent insertion yeast assay for brca missense variant characterization cancer biology therapy vol issue www landesbioscience com cancer biology therapy will not only further define the limits of the assay s usefulness but may also shed some light on the normal function of brca in human cells abstract the brca gene from individuals at risk of breast and ovarian cancers can be screened for the presence of mutations however the cancer association of most alleles carrying missense mutations is unknown thus creating significant problems for genetic counseling to increase our ability to identify cancer associated mutations in brca we set out to use the principles of protein three dimensional structure as well as the correlation between the cancer associated mutations and those that abolish transcriptional activation thirty one of missense mutations of known impact on the transcriptional activation function of brca are readily rationalized in structural terms loss of function mutations involve nonconservative changes in the core of the brca c terminus brct fold or are localized in a groove that presumably forms a binding site involved in the transcriptional activation by brca mutations that do not abolish transcriptional activation are either conservative changes in the core or are on the surface outside of the putative binding site next structure based rules for predicting functional consequences of a given missense mutation were applied to germ line brca variants of unknown cancer association such a structure based approach may be helpful in an integrated effort to identify mutations that predispose individuals to cancer introduction many germ line mutations in the human brca gene are associated with inherited breast and ovarian cancers this information has allowed clinicians and genetic counselors to identify individuals at high risk for developing cancer however the disease association of over missense mutations remains unclear primarily because their relatively low frequency and ethnic specificity limit the usefulness of the population based statistical approaches to identifying cancer causing mutations to address this problem we use here the three dimensional structure of the human brca brct domains to assess the transcriptional activation functions of brca mutants our study is made possible by the recently determined sequences and three dimensional structures of the brca homologs in addition we benefited from prior studies that attempted to rationalize and predict functional effects of mutations in various proteins including those of brca brca is a nuclear protein that activates transcription and facilitates dna damage repair the tandem brct domains at the cooh terminus of brca are involved in several of its functions including modulation of the activity of several transcription factors binding to the rna polymerase ii holoenzyme and activating transcription of a reporter gene when fused to a heterologous dna binding domain importantly cancer associated mutations in the brct domains but not benign polymorphisms inactivate transcriptional activation and binding to rna polymerase ii these observations suggest that abolishing the transcriptional activation function of brca leads to tumor development and provides a genetic framework for characterization of brca brct variants materials and methods the multiple sequence alignment msa of orthologous brca brct domains from seven species including homo sapiens genbank accession number u pan troglodytes af mus musculus u rattus norvegicus af gallus gallus af canis familiaris u and xenopus laevis af was obtained by using program clustalw and contains only one gapped position supplementary fig according to psi blast the latter six sequences are the only sequences in the nonredundant protein sequence database at national center for biotechnology information that have between and sequence identity to the human brca brct domains residues the multiple structure based alignment of the native structures of the brct like domains was obtained by the salign command in modeller supplementary fig it included the experimentally determined structures of the two human brca brct domains protein data bank code jnx refs rat brca brct domains l b ref human p binding protein kzy ref human dna ligase iii\\u03b1 imo ref and human xrcc protein cdz ref structure variability was defined by the root mean square deviation among the superposed c\\u03b1 positions as calculated by the compare command of modeller the purpose of these calculations was to gain insight into the variability of surface exposed residues left panel in fig in conjunction with observed mutation clustering these data may point to putative functional site s on the surface of brct repeats comparative protein structure modeling by satisfaction of spatial restraints implemented in the program modeller was used to produce a three dimensional model for each of the mutants the crystallographic structure of the human wild type brca brct domains was used as the template for modeling the four residues missing in the crystallographic structure and were modeled de novo all of the models are available in the brca model set deposited in our modbase database of comparative protein structure models for the native structure of the human brct tandem repeat and each of the mutant models a number of sequence and structure features were calculated these features were used in the classification tree in fig values for all mutations are given in supplementary tables and buriedness accessible surface area of an amino acid residue was calculated by the program dssp and normalized by the maximum accessible surface area for the corresponding amino acid residue type a residue was considered exposed if its accessible surface area was larger than \\xc5 and if its relative accessible surface area was larger than and buried otherwise a mutation of a more exposed residue is less likely to change the structure and therefore its function functional site the putative functional site was visually assigned to the surface groove that contains the deleterious mutants fig it consists of residue positions with accessible surface area as follows and a mutation in the functional site is more likely to abolish the function residue rigidity a residue isotropic temperature factor b factor was the mean of its atomic b factors as determined by x ray crystallography of the native structure it was also normalized by the mean and standard deviation sd of all residue b factors in the protein chain a mutation of a less rigid residue at a buried position is less likely to change the structure and therefore its function neighborhood rigidity neighborhood of a residue was defined by the residues that have at least one atom within \\xc5 of any atom of the central residue neighborhood b factor of a residue was defined to be the mean value of the b factors of the residues in the neighborhood it was also normalized by the mean and sd of all neighborhood b factors for the neighborhoods of the same size a mutation of a buried residue in a less rigid neighborhood is less likely to change the structure and therefore its function volume change volume change was defined solely by the type of residue substitution a large change in the volume of the amino acid residue type was considered destabilizing to the structure and therefore its function especially when buried at a rigid position charge change charge change was defined solely by the type of residue substitution mutations corresponding to changes in charge at buried positions are more likely to change the structure and therefore its function polarity change polarity change was defined solely by the type of residue substitution amino acid residue types were classified into three classes according to their hydrophobicity embedded image lifwcmvy embedded image patgs embedded image hqrkned mutations corresponding to changes in polarity at buried positions are more likely to change the structure and therefore its function mutation likelihood mutation likelihood of a residue substitution at a given msa position reflects the background substitution probabilities in the blosum matrix as well as the actually observed substitutions at the given position in the msa of the brca brct orthologs supplementary fig refs the scores are centered on zero with negative values corresponding to less likely substitutions unlikely mutations are more likely to change the structure and function phylogenetic entropy sequence variability was defined for every column in the msa of the brca brct orthologs supplementary fig by \\u03a3 p ln p where p is the observed relative frequency of residue type i fig right panel mutations at evolutionarily conserved positions are more likely to change the structure and function helix turn breaker secondary structure assignment was calculated by dssp a helix breaker mutation was defined as any mutation into the glycine or proline residues at a helical position a turn breaker mutation was defined as any replacement of the glycine or proline residues at a turn position visualization and analysis of the mutations were facilitated by the program dino transcriptional assays for testing our predictions were performed as described previously the wild type brca sequence residues fused in frame to the lexa dna binding domain was used as the wild type control and for introducing mutations m r y x m r and w c by site directed mutagenesis egy yeast cells were cotransformed with the constructs and prb which contains a lacz reporter gene under the control of a lexa operator and tested in liquid \\u03b2 galactosidase assays for mammalian transcription assays brca constructs were subcloned in frame to gal dna binding domain in pcdna gal dna binding domain fusion constructs were cotransfected into human t cells with pg luc which contains a firefly luciferase gene under the control of five gal binding sites as a reporter for the assay transfections were normalized with an internal control phrg tk promega which contains a renilla luciferase gene under a constitutive thymidine kinase basal promoter results and discussion we tabulated a total of missense mutations in the human brca brct domains tables and fig these mutations occurred naturally or were obtained via random mutagenesis transcriptional activation function of of mutants has been tested fig rows and eight of the mutants could be tentatively classified either as cancer associated or benign following established genetic criteria ref fig columns and the basis of each classification is outlined in table fig download figureopen in new tabdownload powerpoint fig classification of the missense mutations in the human brca brct domains analyzed in this study the mutants were divided into groups depending on information about their cancer association and transcriptional activation function tables and the mutations at solvent exposed positions are underlined materials and methods view inlineview popup table functionally characterized missense mutations in the human brca brct domains amino acid residues view inlineview popup table prediction of the cancer phenotype of functionally uncharacterized missense mutations in the human brca brct domains for the eight mutations with intact transcriptional activation function there is no evidence of cancer association the frequency in control population is equal or higher than in cancer cases and there is absence of co segregation in at least one family fig row table conversely none of the tested mutations without transcriptional activation function are thought to be benign following the same criteria fig row table ref these observations indicate that tumor suppression by brca correlates with its ability to activate transcription we first rationalized the functional consequences of the tested mutations in terms of their impact on folding and stability of the native structure and in terms of the integrity of binding sites of the brct domains next we used this experience to develop generic rules for prediction of the functional impact of a missense mutation and applied these rules to the remaining functionally uncharacterized mutations listed in fig and table we then tested two of these predictions by functional assays and segregation analysis missense mutations of brca are likely to affect transcriptional activation in either one of the following two ways a when they are exposed to the solvent they may substantially change the structure or chemical nature of functional sites that bind other molecules or b when they are buried in the core they may prevent folding of the brct domains into their native fold or less likely affect only the structure of functional sites the mutations including the fig row and the mutations fig row with and without transcriptional activation function respectively were mapped onto the crystallographic structure of the human brca brct domains ref fig in addition to this structural mapping we produced a three dimensional protein structure model for each mutant and calculated a number of its sequence and structure features materials and methods we used the feature values and their changes between the wild type and the mutant to rationalize the observed effect on transcriptional activation table supplementary table fig download figureopen in new tabdownload powerpoint fig missense mutations on the surface of the human brca brct domains the solvent exposed residues with mutations that are known to affect and not affect the transcriptional activation activity of brca are contoured in red and yellow respectively fig the only remaining exposed mutation of known functional impact p a is on the hidden face of the structure conservation of structure materials and methods supplementary fig and sequence see phylogenetic entropy in materials and methods and supplementary fig are indicated by the intensity of blue on the left and right panels respectively the region in violet on the left panel corresponds to the linker between the two domains the nh and cooh termini of the tandem domains are marked on the right panel the markedly reduced transcriptional activation activity of the mutations fig row is readily explained by their likely disruption of the structure of the brct domain and or its binding sites that are involved in transcriptional activation twenty three of these mutations occur at buried positions and involve amino acid residue types with dissimilar charge hydrophobicity size or secondary structure propensity fig row not underlined thus they are expected to disrupt the structure and therefore its function the remaining six mutations resulting in reduced transcriptional activation are exposed fig row underlined among these six mutations r w and r q replace a conserved residue involved in a salt bridge between the brct repeats another one of the six mutations g r occurs at a position that is absolutely conserved among all available orthologs supplementary fig for the remaining three exposed mutations that abolish transcription we propose that they disrupt interactions with brca ligand s they are located in a groove formed by both brct repeats l p and k e and in the ridge that delimits the groove e g we suggest that this clustering of residues in a single patch on the surface of brca highlights a binding site with integrity that is necessary for transcription moreover this putative binding site is relatively conserved in sequence and structure among the brca homologs fig materials and methods these results are in agreement with in vitro binding studies that determined the binding site of brca to the rna polymerase holoenzyme via rna helicase a nonconservative substitutions of any amino acid residues that comprise the putative binding site are also likely to affect transcriptional activation the existence of such a functional region indicates that the tandem brct repeats comprise one structural and functional unit in accordance with previous conclusions the lack of impact of the eight mutations that do not disrupt transcriptional activation function fig row can also be rationalized in structural terms three of them are exposed on the face away from the putative binding site proposed above whereas the remaining five mutations are conservative changes in the core of the fold table fig five mutations g a s n f i m v and a t in the total set of with impact on the transcriptional activation function that is known could not be readily explained possible reasons for the inability to explain these mutants are discussed in detail in supplementary table because it was straightforward to use the three dimensional structure of the brca domains to rationalize the known functional impact of of the mutations we derived a set of rules for predicting whether a given mutation inactivates transcriptional activation and therefore correlates with predisposition to cancer these rules rely on the sequence and structural features that proved informative in explaining the presence or absence of a functional impact of the tested mutations for example structural neighborhood around a mutated residue can be rigid and a mutation can introduce a large change over a in the side chain volume such a mutation is likely to destabilize or distort local conformation thus having an effect on the function of the protein in contrast a flexible structural neighborhood is expected to accommodate such a substitution without major perturbations in the structure moreover our rules allow for the possibility that very large changes in the side chain volume over a can have a significant effect even in a flexible neighborhood the rules were encoded hierarchically in a classification tree scheme fig and are implemented on the web fig download figureopen in new tabdownload powerpoint fig classification tree for annotating a given missense mutation in the human brca brct domains the classification begins at the top start and ends at one of the or leaves the and indicate mutants with intact benign and abolished transcriptional activation function cancer associated respectively this classification tree correctly rationalizes of the functionally characterized variants table the tree was also applied to mutants that have not yet been characterized by experiment table see materials and methods and supplementary material for a detailed description of the features using the classification tree we predicted the functional consequences of the missense mutations in the brct domains that presently lack definitive information about cancer association breast cancer information core database table supplementary table our prediction algorithm assigned functional impact to mutants whereas were predicted to be benign to validate our predictions we tested prospectively two mutants fig a and b for which segregation data were available the w c mutation was found in a family with three generations affected where women were diagnosed with either early onset years old breast cancer or ovarian cancer it was present in four of four women with cancer and was absent in three of three at risk women without cancer the m r mutation also segregated with disease in a large family with several generations affected fig c both m r and w c displayed the loss of function phenotype and segregated with disease in agreement with our predictions fig fig download figureopen in new tabdownload powerpoint fig transcription assays and segregation analysis confirm predictions for mutations m r and w c structures of the reporter plasmid are depicted on top of panels a and b cancer associated mutations m r and y x were used as controls a transcriptional activity in yeast cells b transcriptional activity in human cells c pedigree with segregation analysis for the m r mutant arrowhead indicates the proband the presence m r or absence neg of m r mutation in the germ line of tested individuals is indicated site of tumor and age of diagnosis is also indicated wt wild type to illustrate prediction by the classification tree we used the mutation m r that occurs at a buried position in a relatively rigid neighborhood table supplementary table mutations at buried and rigid locations are tolerable only if changes in the residue volume charge and polarity are sufficiently small a volume change of more than one methyl group \\xc5 cannot be accepted however the arginine volume is only \\xc5 larger than that of methionine therefore the condition on the maximal allowed volume change is satisfied at the next checkpoint the side chain charge is considered because m r introduces a positive charge at a buried position the mutation is considered destabilizing resulting into the prediction of a loss of function the tumor suppressor function of brca is likely to depend at least partially on its transcriptional activity as suggested previously and supported by the mutation set in table therefore the ability to predict which mutations do and do not abolish the transcriptional activation function of brca can be useful in predicting their cancer association our analysis based primarily on the considerations of protein three dimensional structure leads to a prediction of cancer association of the mutations that cannot be easily characterized by other approaches such as population based statistical genetics although all but two of the present predictions are unverified and thus cannot be used as a sole tool for risk assessment and genetic counseling this structure based approach may be helpful in an integrated effort to identify mutations that predispose an individual to breast and ovarian cancers mutations in brca and brca account for the majority of hereditary breast and ovarian cancers and therefore sequence analysis of both genes is routinely conducted in patients with early onset breast cancer besides mutations that clearly abolish protein function or are known to increase cancer risk a large number of sequence variants of uncertain signifi cance vus have been identifi ed although several functional assays for brca vuss have been described thus far it has not been possible to conduct a high throughput analysis in the context of the full length protein we have developed a relatively fast and easy cdna based functional assay to classify brca vuss based on their ability to functionally complement brca defi cient mouse embryonic stem cells using this assay we have analyzed unclassifi ed brca missense mutants for which all predicted pathogenic variants are confi ned to the brca ring and brct domains significance brca vuss are frequently found in patients with hereditary breast or ovarian cancer and present a serious problem for clinical geneticists this article describes the generation validation and application of a reliable high throughput assay for the functional classifi cation of brca sequence variants of uncertain signifi cance cancer discov aacr authors affi liations division of molecular pathology and cancer genomics centre and department of pathology the netherlands cancer institute amsterdam the netherlands and taconicartemis gmbh cologne germany note supplementary data for this article are available at cancer discovery online http cancerdiscovery aacrjournals org current address for c klijn genentech inc south san francisco california corresponding author jos jonkers the netherlands cancer institute plesmanlaan amsterdam cx the netherlands phone fax e mail j jonkers nki nl doi cd american association for cancer research introduction germline loss of function mutations in brca and brca are known to result in an approximately tenfold increased lifetime risk of developing breast or ovarian cancer thus far no other genes have been identifi ed with such a strong link to hereditary breast and ovarian cancer hboc and in the past decades many women have been screened for germline mutations in brca or brca this has resulted in the identifi cation of numerous pathogenic mutations as well as a large number of sequence variants for which the clinical relevance is not clear in the most recent publication of the enigma evidence based network for the interpretation of germline mutant alleles group an international consortium for the evaluation of brca or brca sequence variants a total of unique brca variants is mentioned but this number may increase even further because of the implementation of high throughput sequencing methods of the brca variants of uncertain signifi cance vus are nontruncating exonic mutations that may affect protein function or mrna splicing but for which there is not enough linkage information to indicate whether they are pathogenic to aid genetic counseling of individuals with brca vus both genetic and functional classifi cation methods have been developed genetic analysis of brca or brca vuss relies on cosegregation with disease cooccurrence with known pathogenic mutations and family history of cancer these data have been integrated into computational models to calculate the likelihood that a vus is disease causing additional in silico analysis of the evolutionary conservation of the amino acids affected by the mutation and the predicted impact of the mutant amino acids on protein folding are also implemented in such models functional assays do not rely on preexisting data but directly test the effect of brca vuss on known functions of the encoded proteins although this may seem relatively straightforward it can be diffi cult to extrapolate data from functional assays into cancer risks for patients for instance in most assays only part of the brca protein is analyzed in addition some of the more elegant assays are technically demanding and not suitable to analyze large numbers of mutations we reasoned that a good functional assay should fulfi ll three basic requirements i it should investigate the biologic effects of a brca vus in the context of the full length protein ii it should be conducted under normal physiologic conditions in a noncancerous cell type and iii it should be based on a highly standardized and reproducible protocol these considerations led us to develop a functional assay based on physiologic expression of full length human brca cdna in mouse embryonic stem cells that are genetically engineered to allow conditional deletion of endogenous brca mutant brca cdnas are generated using site directed mutagenesis sdm and introduced in a defi ned genomic locus of mouse embryonic stem cells by recombinase mediated cassette exchange rmce in this way we have analyzed brca variants for their effects on cell proliferation and drug sensitivity including clinically relevant vuss results generation of mouse embryonic stem cells expressing brca sequence variants although brca defi cient tumor cells proliferate rapidly in situ loss of brca in normal cells leads to a severe proliferation defect we decided to make use of this phenotype for the functional analysis of brca variants in downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd cancer discovery october www aacrjournals org research article bouwman et al brca selectable conditional knockout sco mouse embryonic stem cells these cells carry one brca null allele and a selectable conditional brca sco allele which contains in addition to loxp sites around exons a split puromycin resistance marker that is activated upon cre mediated deletion of exons they also contain a creert allele in the rosa locus which allows for controlled activation of cre via administration of hydroxy tamoxifen oht thus brca defi cient embryonic stem cells can be easily obtained via oht induced inactivation of the brca sco allele and subsequent selection for puromycin resistance fig ref to allow effi cient integration of human brca variants in one specifi c genomic locus we supplied the other rosa allele with f and frt recombination sites for rmce by the site specifi c recombinase flp fig and supplementary fig s ref cells that have undergone successful rmce can be selected because they express a truncated neomycin selection marker under control of the endogenous rosa promoter which further increases targeting effi ciency flpmediated recombination ensures single copy integration of brca expression constructs at the same rosa locus thus avoiding position effect variegation and copy number dependent differences in expression rmce vectors were equipped with a human brca cdna expression construct which was modifi ed using sdm to introduce defi ned mutations in brca the focus of our analysis was on dutch and belgian vuss that were found in families with hboc we also included a number of variants figure schematic overview of the rmce procedure in r creert rmce brca sco \\u0394 embryonic stem cells before the introduction of a human brca cdna r creert rmce brca sco \\u0394 embryonic stem cells are mouse brca profi cient and sensitive to both neomycin and puromycin targeting of singlecopy human brca cdna variants to the rosa locus by flp rmce results in expression of human brca and neomycin resistance addition of oht leads to creert mediated deletion of mouse brca exons and resulting in loss of mouse brca protein and concomitant expression of puromycin from the phosphoglycerate kinase pgk promoter this enables selection of mouse brca defi cient human brca expressing r creert hbrca brca \\u0394 \\u0394 embryonic stem cells that can be used in functional complementation assays r creert rmce brca sco \\u0394 r creert hbrca brca sco \\u0394 r creert hbrca brca \\u0394 \\u0394 introduction of human brca by rmce sa sa brca sco brca \\u0394 r creert r rmce brca \\u0394 brca \\u0394 r creert r hbrca brca sco brca sco brca \\u0394 r creert r rmce brca \\u0394 r creert r hbrca creert creert creert creert puro puro puro pgk pgk pgk puro pgk zsgreen hyg flp zsgreen hyg flp sa sa neor selection hbrca mbrca neos puros hbrca mbrca neor puros hbrca mbrca neor puror pa hbrca hbrca ef \\u03b1 ef \\u03b1 \\u0394 neo \\u0394 neo \\u0394 neo ef \\u03b1 hbrca cre mediated deletion of mouse brca by oht addition puror selection sa sa sa sa downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd october cancer discovery functional analysis of brca sequence variants research article that were previously classifi ed using functional assays or a multifactorial likelihood model as well as m v l p and m t which have been reported to attenuate the interaction between brca and palb to allow validation of our functional complementation assay we included a series of eight brca variants that are known to be deleterious or neutral according to the breast cancer information core database bic http research nhgri nih gov bic supplementary table s these controls include the well known pathogenic brca founder mutations delag and insc and the neutral polymorphisms y c r c and e k the bic designation is supported by the classifi cation according to the international agency for research on cancer http iarc fr which includes the align gvgd score that indicates biophysical and evolutionary alterations http agvgd iarc fr supplementary table s refs align gvgd scores vary between c likely neutral and c likely deleterious sequence verifi ed constructs that contained the intended vus but no other brca mutations were transfected into r creert rmce brca sco \\u0394 embryonic stem cells to undergo rmce fig neomycin resistant clones were pooled and rmce was confi rmed by pcr analysis subsequently protein expression of the human brca variants was analyzed by western blot analysis with a human brca specifi c antibody fig and supplementary fig s comparison with brca protein levels in embryonic stem cells expressing human brca from a bacterial artifi cial chromosome bac which is known to rescue embryonic lethality of brca null mice showed that transcription of the human brca cdna from the ef a gene promoter results in physiologic levels of brca protein supplementary fig s most brca variants were expressed at equal levels allowing comparison of their functional activities a number of c terminal brca mutants showed low levels of expression supplementary table s supplementary fig s real time rt pcr analysis showed that decreased protein expression was not caused by decreased mrna expression supplementary fig s suggesting that the low abundance of brca protein results from posttranslational events in fact for most of these variants mutation associated protein instability has already been documented in previous publications refs supplementary table s functional complementation assay of brca sequence variants in mouse brca null embryonic stem cells as a fi rst functional test we assayed the ability of brca variants to restore the proliferation defect of switched r creert rmce brca sco \\u0394 embryonic stem cells fig expression of endogenous mouse brca was shut off through overnight induction of cre activity by oht and days figure workfl ow for the functional classifi cation of brca sequence variants in brca null embryonic stem es cells outline of the generation of mouse brca defi cient embryonic stem cells expressing human brca variants and functional complementation assays indicated are the experimental steps and the time it takes one person to analyze mutants introduction of brca variants in r creert rmce brca sco d es cells brca construct g selection of clones with correct rmce brca expression analysis sdm and sequencing rmce in es cells x oht puromycin functional assays functional assays cre mediated deletion of brca sco allele proliferation and cisplatin sensitivity assays selection of brca \\u0394sco \\u0394 es cells timecourse for the analysis of brca variants weeks sdm and rmce functional assays downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd cancer discovery october www aacrjournals org research article bouwman et al after switching cells were plated in well plates for proliferation assays and were analyzed using sulphorhodamine b staining for each group of mutants tested we included positive and negative controls consisting of embryonic stem cells containing respectively wild type human brca cdna and an empty rmce vector brca variants were evaluated on their ability to support growth compared with these controls supplementary tables s and s although we tested mutations in the central domain encoded by exon aa brca variants that were unable to rescue the proliferation defect of brca null mouse embryonic stem cells to brca wild type levels were confi ned to the conserved n and c terminal domains of brca cisplatin sensitivity assay for classifi cation of brca variants although the ability of brca variants to support proliferation appears to be indicative of vus function results were not always clear cut supplementary tables s and s brca is known to be important for dna interstrand crosslink icl repair through mechanisms that are both dependent and independent of its function in homologous recombination hr the role of brca in icl repair is stressed by the occurrence of genetic reversion mutations restoring brca protein expression in platinum resistant ovarian tumors in brca mutation carriers we therefore decided to conduct a well based cisplatin sensitivity assay to allow a more stringent evaluation of brca vus functionality half maximal inhibitory concentrations ic of cisplatin were determined using a resazurin cell viability assay and brca variants were again classifi ed in comparison to wildtype brca and an empty rmce vector to obtain corrected cisplatin ic values we fi tted a log logistic curve constrained at and we excluded a fi t if the residual squared error rse exceeded we then applied a bayesian predictor to classify brca vus as pathogenic or benign most variants that showed less than wild type activity in the proliferation assay also scored as functionally impaired in the cisplatin sensitivity assay and were classifi ed as deleterious table table functional classifi cation of brca vus based on cisplatin response variant dna change type of mutation a classifi cation s f c c t vus not clear m t c t c vus deleterious delag c _ delag deleterious control deleterious b k q c a c vus neutral c g c t g deleterious control deleterious c g c t g vus deleterious b d y c g t vus neutral y c c a g neutral control neutral n k c c a vus neutral p h c c a vus neutral l f c g c vus neutral l p c t c vus neutral r w c c t vus neutral s y c c a vus neutral v i c g a vus neutral t m c c t vus neutral d v c a t vus neutral l v c t g vus neutral v l c g c vus neutral s a c a g c g c artifi cial h not clear l r c r e h c t g c t c c g c c g t artifi cial neutral l f c c t vus neutral d n c g a vus neutral p l c c t vus neutral n y c a t vus neutral t k c c a vus neutral continued downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd october cancer discovery functional analysis of brca sequence variants research article continued variant dna change type of mutation a classifi cation r q c g a vus neutral y h c t c vus neutral r c c c t neutral control neutral s a c t g artifi cial neutral m i c g a vus neutral e a c a c vus neutral s n c g a vus neutral k del c _ delaag vus neutral s g c a g vus neutral e k c g a vus neutral n k c t g vus neutral e k c g a neutral control neutral l s c t c vus neutral e v c a t vus neutral s del c _ delttc vus neutral s r c t a vus neutral e k c g a vus neutral v i c g a vus neutral m v c a g vus neutral l p c t c vus neutral b m t c t c vus neutral r g c c g vus neutral s g c a g vus neutral s c c a t vus neutral s a c t g artifi cial neutral v m c g a vus neutral r p c g c vus neutral m t c t c vus neutral s p c t c vus not clear s f c c t vus not clear m i c g a vus neutral s f c c t vus deleterious b h r c a g vus deleterious h q c t a vus deleterious v del c _ delgtt vus deleterious t i c c t vus not clear r w c c t deleterious control deleterious b r q c g a vus deleterious b g e c g a vus deleterious b g a c g c vus neutral a e c c a deleterious control deleterious w c c g t vus deleterious t a c a g vus neutral e k c g a vus not clear table functional classifi cation of brca vus based on cisplatin response continued downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd cancer discovery october www aacrjournals org research article bouwman et al fig variants showing increased cisplatin sensitivity were tested at least twice before they were classifi ed the positive and negative controls classifi ed as expected although the known pathogenic truncation mutation insc scored as neutral in one of three assays stressing the need for repeat experiments also brca variants that were previously tested in other assays conducted as expected it should be noted that the artifi cial variants s a and s a rescued proliferation and cisplatin responses of mouse brca defi cient embryonic stem cells in bac complementation assays but are predicted to be deleterious based on their effects on embryonic stem cell differentiation and their response to \\u03b3 irradiation respectively the v d mutation scored as a neutral variant in our assay which is in line with most published data except for the results of an embryonic stem cell based bac complementation assay again all brca mutations that were classifi ed as deleterious were confi ned to regions encoding the conserved n and c terminal domains fig despite the observation that deletion of the central region encoded by exon leads to genetic instability in mice remarkably the three mutations that diminish the interaction between brca and palb just c terminal of the region encoded by exon had no effect on cisplatin sensitivity in this assay of note there was in general good correlation between our classifi cation and the align gvgd score supplementary table s notable exceptions were the neutral control variant r c which validated our assay but scored as likely deleterious c using align gvgd and g v which scored as likely neutral c by align gvgd but was classi fi ed as deleterious in our assay our assay also classifi ed some variants for which the align gvgd scores were less clear emphasizing the usefulness of functional assays to complement in silico analysis we also analyzed the possible effects of all brca variants on mrna splicing which may have deleterious consequences but cannot be assessed in our cdna based assay the predictive value for exonic variants outside the consensus splice sites is questionable but four missense mutations were present in existing splice sites supplementary table s although c g t encoding w c was deleterious three of these variants were classifi ed as neutral in our assay one of them c g c encoding l f had an increased probability to damage the splice donor site of exon another variant c c t encoding t i might also affect splicing but no defect was measured in blood samples of mutation carriers the third exonic splice site mutation that was classifi ed as neutral in our cisplatin sensitivity assay c a c k q is not predicted to lead to aberrant splicing results from proliferation and cisplatin sensitivity assays correlate with hr activity of brca variants brca and brca are involved in dna repair via hr together with non homologous end joining nhej hr forms the cellular defense against dna double strand breaks dsb a severe type of dna damage that is lethal if unrepaired although hr is essentially error free nhej is error prone and therefore defects in hr are known to lead to genomic instability although it is not clear whether other variant dna change type of mutation a classifi cation v a c t c vus not clear b d g c a g vus deleterious d v c a t vus deleterious h q c c g vus not clear r t c g c vus not clear b insc c dupc deleterious control not clear b l p c t c vus deleterious b c s c g c vus neutral g v c g t vus deleterious b w c c g t vus neutral a t c g a vus deleterious e d c g c vus neutral v d c t a vus neutral p r c c g vus neutral w r c t c vus deleterious h l c a t vus neutral a type of mutation indicates if a variant is a vus according to the bic database http research nhgri nih gov bic b functionally impaired in the dr gfp and or combined parp inhibitor cisplatin sensitivity assay table functional classifi cation of brca vus based on cisplatin response continued downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd october cancer discovery functional analysis of brca sequence variants research article figure waterfall chart of cisplatin ic values normalized for wild type human brca and empty rmce vector controls corrected cisplatin ic values for all tested human brca variants and controls including repeat experiments brca variants are classifi ed as functionally impaired or neutral when the corrected ic values are similar to either the empty rmce vector or the wild type human brca controls p corrected cisplatin ic a u mutant hbrca wild type hbrca empty rmce vector positive control negative control figure predicted pathogenic brca amino acid substitutions are confi ned to the evolutionarily conserved n and c terminal domains schematic representation of the brca protein with the positions of variants classifi ed as neutral green or deleterious red indicated positive and negative controls are depicted by open pinheads and the known deleterious truncation mutations delag n terminal and insc c terminal are marked in blue the amino acid sequences of the evolutionarily conserved ring n terminal and brct c terminal domains are specifi ed to show the exact positions of deleterious red and neutral green variants the amino acid substitutions or nonsense x mutations of positive and negative controls are encircled ring domain brct domains deleterious vus neutral vus deleterious truncation mutant deleterious control variant neutral control variant mdlsalrveevqnvinamqkilecpiclel vnkrmsmvvsgltpeefmlvykfarkhhitltnliteetthvvmktdaefvcertlkyfl ikepvstkcdhifckfcmlkllnqkkgpsq giaggkwvvsyfwvtqsikerkmlnehdfevrgdvvngrnhqgpkraresqdrkifrgle cplcknditkrslqestrfsqlveellkii cafqldtgleyansynfak iccygpftnmptdqlewmvqlcgasvvkelssftlgtgvhpivvvqpdawtedngfhaig qmceapvvtrewvldsvalyqcqeldtylipqiphshy t q e c a x y c gg f r q w q i e a v r l s c d dr t v p g x downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd cancer discovery october www aacrjournals org research article bouwman et al functions of brca are also important for tumor suppression its role in hr is likely to be relevant therefore we used the direct repeat dr gfp assay to measure the effects on hr for a subset of our brca vuss a number of predicted pathogenic variants and controls were shuttled into r creert rmce brca sco \\u0394 pim dr gfp wt embryonic stem cells carrying the dr gfp reporter in the pim locus expression of endogenous mouse brca was switched off and cells were transfected with a plasmid encoding the i scei meganuclease as well as an mcherry fl uorescent marker to control for transfection effi ciency repair of i scei induced dna dsbs in dr gfp via hr leads to expression of gfp which can be monitored by fl ow cytometry all but one of the predicted pathogenic variants tested resulted in defective hr thereby confi rming our functional classifi cation fig the only predicted deleterious variant that did not signifi cantly differ from wild type brca r q seemed to support intermediate levels of hr activity parp inhibitor sensitivity assay for classifi cation of brca variants with intermediate activity deleterious effects of variants with intermediate or partial activity may escape detection in certain functional assays although cisplatin sensitivity assays allow robust and figure brca sequence variants classifi ed as pathogenic do not restore hr r creert hbrca brca sco \\u0394 pim dr gfp wt embryonic stem cells carrying the dr gfp reporter gene in the pim locus and mutant brca or controls in the rosa locus were switched using oht and transfected with a vector expressing i scei and mcherry transfected cells were analyzed for gfp expression as a measure of hr activity expression of brca wild type wt cdna resulted in increased hr compared to the empty rmce vector vector control error bars indicate the sd between the results of three independent i scei transfections signifi cantly decreased hr activity compared with the wild type control green line is indicated gfp positive cells ns ns ns ns wt vector delag c g v l s f s f r q r w g a g e v a r t insc l p g v reproducible classifi cation of several functionally impaired brca variants assays using other compounds may have additional value it is known that brca and brca defi cient cells are extremely sensitive to parp inhibition leading to a larger dynamic range between brca defi cient cells and isogenic brca profi cient controls than for cisplatin we therefore tested complementation of parp inhibitor sensitivity for a number of brca mutants and the brca wild type control given the unexpected neutral effects of the m v l p and m t mutations in the palb interaction domain we decided to include these variants in this series as well as the r q and v a variants that have recently been shown to confer intermediate breast and ovarian cancer risk to allow direct comparison of results from different assays we repeated the cisplatin sensitivity and proliferation assays in parallel to the olaparib sensitivity assay our results indicate that brca wild type and empty vector controls indeed show a larger difference in sensitivity for olaparib than for cisplatin fig however this increase in dynamic range is accompanied by an increased variation between repeat experiments for brca profi cient samples nevertheless the functional defect of the r q and v a mutations becomes more evident and there also downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd october cancer discovery functional analysis of brca sequence variants research article figure parp inhibitor sensitivity assay of brca sequence variants r creert hbrca brca sco \\u0394 or r creert rmce brca sco \\u0394 embryonic stem cells carrying mutant brca brca wild type wt or empty rmce vector vector controls in the rosa locus were switched using oht and assayed for sensitivity to cisplatin or the parp inhibitor olaparib the cytotoxicity assays were conducted in parallel and data were normalized to the average of the wild type controls error bars indicate the sd between the results of biologic triplicates for which the cells were independently switched signifi cant deviation from the average ic values of the wild type control green line is indicated ic relative to wt controls cisplatin olaparib ns ns ns ns wt vector m v l p m t r q v a seems to be a less than wild type response for the palb interaction mutant l p in the concurrent proliferation analysis r q and v a both show an intermediate functional defect supplementary fig s also the m t mutation seems to affect the response to parp inhibition but the difference with brca wild type is not signifi cant interestingly the l p and m t variants have previously been shown to be more defective than m v in a gene conversion assay it should also be noted that in contrast with the large scale classifi cation experiments the cisplatin sensitivity assay conducted in parallel to the parp inhibitor assay identifi ed signifi cant functional defects for l p and v a discussion over the past few years several functional assays for classifi cation of brca vuss have been developed several of these assays are restricted to functions of the brca protein that reside in the evolutionarily conserved ring or brct domains examples include in vitro transactivation assays for brct peptides and measurement of ubiquitin ligase activity for protein fragments encompassing the n terminal ring domain other assays were designed to evaluate the functions of full length mutant brca protein either by monitoring general effects on proliferation or response to dna damage or by directly focusing on the role of brca in dna repair via hr because brca vuss are not restricted to regions encoding the n or c terminal domains and given the observation that interaction between these domains is required for recruitment of brca to damaged dna functional assays for the full length protein would be ideal in principle such assays can be conducted in cell lines derived from brca mutated tumors but there are indications that the outcome of assays for brca function depends on the cellular context as absence of brca leads to loss of cellular viability it is thought that additional mutations are required for brca associated tumorigenesis for example loss of p alleviates the consequences of brca defi ciency both in vitro and in vivo and is common in brca defi cient tumors also depletion of bp is known to suppress the defects caused by brca defi ciency therefore aberrations in brca defi cient tumor cells may mask functional defects of brca vuss we reasoned that assays in normal cells that can be depleted from endogenous brca expression are most likely to reveal functional defects of brca vuss the usefulness of this approach has been previously shown by downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd cancer discovery october www aacrjournals org research article bouwman et al bac complementation assays in mouse embryonic stem cells however mutagenesis of large bac clones by recombineering and functional complementation of cells with these mutant constructs is time consuming and technically demanding and therefore cannot be conducted in a high throughput setting we therefore set out to develop a cdna based functional complementation assay in mouse embryonic stem cells that is easier to control and scale up and more suitable for routine functional classifi cation of brca sequence variants one advantage of bac transgenics in complementation assays is that genes are expressed at physiologically relevant levels this is the result of low copy number integrations and the presence of natural regulatory elements required for proper gene expression we decided to use rmce to allow single copy integration of brca cdnas at one specifi c genomic locus the use of rmce effectively prevents multiple or partial integrations concatemers and position effect variegation as a result all variants are expressed at equal levels moreover transcription of brca cdnas from the ef a gene promoter results in physiologic levels of brca protein comparable with those observed in embryonic stem cells stably transfected with a bac containing the human brca locus indeed the wild type human brca cdna was able to complement brca null embryonic stem cells in cellular proliferation drug sensitivity and hr assays our rmce strategy uses brca cdna constructs in which mutations can be swiftly introduced using sdm enabling a higher throughput than introduction of mutations via bac recombineering in addition the introduction of brca cdnas via rmce obviates the need to analyze multiple embryonic stem cell clones for correct integration and expression as a proof of principle we used our functional complementation assay to analyze exonic brca vuss that were identifi ed in families with hboc in the netherlands and belgium as well as a set of previously analyzed brca variants a number of mutations resulted in reduced brca protein levels most likely because of structural destabilization in all cases this led to diminished capacity for functional complementation also several variants that gave rise to normal brca protein levels were unable to rescue the proliferation defect and cisplatin sensitivity of brca null embryonic stem cells as brca loss of function mutations are associated with increased cancer risk variants that score as functionally impaired in our embryonic stem cell assay system may be causally involved in tumor formation this notion is supported by the fact that seven of eight known pathogenic or neutral control variants in our validation series were correctly classifi ed by the cisplatin sensitivity assay the pathogenic insc truncation mutation could not be classifi ed because it scored as neutral in one of three transfection series this was probably due to technical reasons as the insc mutation did not restore hr activity in brca defi cient embryonic stem cells in contrast with r q which was recently shown to confer intermediate risk of hboc our assay system yielded ambivalent results for nine other variants s f s a s p s f t i v a e k h q and r t t i and e k were classifi ed as functionally impaired in only one cisplatin sensitivity test whereas values from repeat experiments could not be taken into account because of rse values above s f s p s f v a h q and r t were differently classifi ed in repeat experiments which may refl ect technical fl aws or intermediate activity of these variants of note the v a mutation was recently identifi ed as a pathogenic variant with hypomorphic activity in dna repair although we did not measure signifi cant hr activity of v a in a dr gfp gene conversion assay intermediate activity of this variant is supported by the results of the proliferation assays and additional cisplatin sensitivity assays s f showed hr activity similar to wild type brca whereas r t was hr defi cient s a is an artifi cial mutation of a brca phosphorylation site that was able to support proliferation and resistance to dna damage in an embryonic stem cell based bac complementation assay however s a complemented embryonic stem cells did show increased apoptosis when cultured in embryoid bodies indicating a partial defect that might explain the ambivalent results for this mutation in our assays our results show that brca variants should ideally be assayed in triplicate to avoid misclassifi cation this also applies to brca vus that we classifi ed as neutral most of which were tested only once because in our current study we focused on variants that showed functional impairment the striking restriction of unambiguously predicted pathogenic mutants to the terminal ring and brct domains suggests that some plasticity is allowed for the central domain of brca however our dataset is still limited and more experiments are required to gain insight into the function of this domain our cdna based system allows for several additional functional assays that have been described previously for bac transgenic embryonic stem cells these include assays for defects during in vitro and in vivo embryonic stem cell differentiation but also treatments with other cytotoxic agents as a proof of principle we investigated the activity of a number of mutants in the response to the parp inhibitor olaparib these included m v l and m t which were previously shown to impair palb binding and have a negative effect on brca function although the differences in sensitivity of brca defi cient versus brca profi cient embryonic stem cells are larger for olaparib than for cisplatin increased variation between repeat experiments allowed us to identify functional defects only for l and not for the other two variants in the palb interaction domain m v and m t however the r q and v a variants which are known to have hypomorphic activity clearly showed a defect in the response to olaparib interestingly both mutations are in the brct domain and it has recently been shown that mutation of this domain especially confers sensitivity to parp inhibition together our results show that parp inhibitor sensitivity assays may have added value especially for the classifi cation of brca vuss with intermediate phenotypes platinum drugs and parp inhibitors are selectively toxic to brca defi cient cells because they target hr defi ciency although the role of brca in hr is thought to be essential for maintaining genomic integrity and preventing accumulation of oncogenic mutations other activities may also contribute to its tumor suppression function these activities downloaded from cancerdiscovery aacrjournals org on may american association for cancer research published onlinefirst july doi cd october cancer discovery functional analysis of brca sequence variants research article may include the hr independent role for brca in icl repair which has been attributed to the facilitation of fancd accumulation at cross linked dna nevertheless we observed a good correlation between the results of the cisplatin sensitivity assay and the results of the dr gfp hr assay all brca variants that failed to restore the cisplatin response in brca null embryonic stem cells were also defective in catalyzing gene conversion thereby confi rming our functional classifi cation it will be interesting to see if this holds true for all brca variants or whether there are also pathogenic mutations that have no effect on hr as with any other in vitro approach our functional complementation assay system might still fail to identify all pathogenic variants because it does not necessarily recapitulate all aspects of brca function in vivo a limitation of our cdna based assay is also that it cannot be used to investigate effects on mrna splicing although algorithms have been designed to predict possible splice defects the consequences of mutations outside of the consensus splice sites especially require functional validation experiments for this purpose bac complementation assays minigene based splicing assays or brca transcript analysis of patient blood samples may be instrumental however transacting factors also affect splicing and these may be tissue specifi c a possible solution would be to determine the presence of brca splice variants in tumor tissue from brca vus carriers and use this information to generate a cdna construct for analysis of the functional consequences to evaluate brca vuss there remains a need for multifactorial models that combine results from functional assays and in silico analyses with genetic evidence and other information from mutation carriers this also includes dna copy number data from tumors from brca vus carriers as it is known that brca associated breast tumors show distinct genomic aberrations our functional assay system does however provide a robust and easily implementable tool for the functional characterization of large numbers of brca vuss within the context of the full length protein it is our hope that our assay system will fi nd its way to clinical genetics laboratories where it can be used to aid genetic counseling ideally these tests should be coordinated on an international level and in close collaboration with the enigma consortium genetic screening of the breast and ovarian cancer susceptibility gene brca has uncovered a large number of variants of uncertain clinical significance here we use biochemical and cellbased transcriptional assays to assess the structural and functional defects associated with a large set of distinct brca missense variants within the essential brct domain of the brca protein that have been documented in individuals with a family history of breast or ovarian cancer in the first method we used limited proteolysis to assess the protein folding stability of each of the mutants compared with the wild type in the second method we used a phosphopeptide pull down assay to assess the ability of each of the variants to specifically interact with a peptide containing a pser x x phe motif a known functional target of the brca brct domain finally we used transcriptional assays to assess the ability of each brct variant to act as a transcriptional activation domain in human cells through a correlation of the assay results with available family history and clinical data we define limits to predict the disease risk associated with each variant forty two of the variants show little effect on function and are likely to represent variants with little or no clinical significance display a clear functional effect and are likely to represent pathogenic variants and the remaining variants display intermediate activities the excellent agreement between the structure function effects of these mutations and available clinical data supports the notion that functional and structure information can be useful in the development of models to assess cancer risk introduction since the cloning of brca in a large effort has been made to sequence the brca genes of women who are at increased risk for early onset breast and ovarian cancers corresponding author j n mark glover or alvaro n a monteiro department of biochemistry university of alberta edmonton alberta t g h canada phone fax mark glover ualberta ca or alvaro monteiro moffitt org reprints and subscriptions to order reprints of this article or to subscribe to the journal contact the aacr publications department at pubs aacr org note supplementary data for this article are available at cancer research online http cancerres aacrjournals org disclosure of potential conflicts of interest no potential conflicts of interest were disclosed nih public access author manuscript cancer res author manuscript available in pmc june published in final edited form as cancer res june doi can nih pa author manuscript nih pa author manuscript nih pa author manuscript this effort has revealed several mutations that are strongly linked to cancer unfortunately a large number of rare brca variants have been uncovered in the human population for which risk assessment has been problematic due to a lack of informative pedigree data linking each mutation with disease risk brca encodes a large amino acid nuclear protein that plays a critical role in the response of cells to genotoxic stress brca is thought to act as an essential mediator protein in dna damage induced nuclear signaling events in which it interacts with phosphorylated partner proteins such as the dna helicase bach the nuclease ctip and another signaling protein abraxas to relay signals generated from chromatin surrounding the damage to downstream targets such as dna repair proteins and factors involved in cell cycle regulation interactions with phosphorylated partner proteins are mediated by a tandem pair of repeats at the cooh terminus of brca termed brct repeats the brct repeat region can also act as a transcriptional activation domain when linked to a sequence specific dna binding module and this activity may contribute to the ability of brca to regulate the expression of genes such as p and gadd and modulate the activity of other transcription factors such as p and er the nh terminal region of the protein contains a ring domain which forms a heterodimeric complex with the ring domain of bard to form a ubiquitin ligase cancer associated mutations tend to cluster in the ring and brct repeat regions showing the critical role of these domains in brca tumor suppression the brca brct domain specifically interacts with phosphorylated protein targets containing the motif pser x x phe structural studies have revealed that recognition involves a conserved phosphoserine recognition pocket in the nh terminal brct repeat composed of ser gly and lys which each supply ligands to recognize the phosphate the phenylalanine residue at the position of the peptide target is recognized by a largely hydrophobic groove formed at the interface between the repeats phosphopeptide recognition is likely a common function of brct domains in many proteins associated with dna damage mediated signal transduction for example the phosphorylated form of the histone variant h ax which serves as a critical chromatin mark of dna double strand breaks is specifically bound by the tandem brct repeats of mdc to initiate dna doublestrand break signaling several studies have attempted to predict the cancer risks associated with unclassified missense variants in brca through bioinformatics analysis based on multiple sequence alignment data and protein structure prediction as well as analyses of the clinical and family history data available for some of these mutations these studies indicate that the ring and brct domains which are the most highly conserved regions of brca likely contain the vast majority of cancer associated mutations here we use in vitro protein folding phosphopeptide binding and cell based transcriptional assays to assess the structural and functional consequences of a set of distinct missense variants that have been uncovered in the brca brct domain in the human population our results indicate that nearly one half of the brct missense variants result in significant structural and or functional defects that are likely associated with an increased cancer risk materials and methods proteolysis assays plm plasmid \\u03bcg encoding the brct variants was used directly as template for protein synthesis reactions with the tnt quick in vitro transcription translation system promega both wild type and mutant proteins were produced by in vitro transcription translation because many mutants could not be expressed in soluble form in escherichia coli the stability of the protein fold for the brca brct mutants was tested using lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript proteolysis assays as previously reported the measurements were normalized against the internal background and the control measurements taken in the same day the missense variants were tested in triplicate phosphopeptide binding assays the brct variants were translated and labeled as in proteolysis assays the ability of the brca brct mutants to bind pser containing peptides was tested using peptide binding assays as previously reported bead immobilized peptide affinity resin was prepared by incubating a fold molar excess of a biotinylated phosphopeptide srstpsptfnk and the corresponding unphosphorylated peptide srstsptfnk with streptavidin agarose beads sigmaaldrich seven mutants with the wild type brca brct as a positive control and m r and a e as negative controls were examined at a time the intensities of the bands were measured and the backgrounds were corrected using imagequant amersham biosciences and the measurements were normalized against the wild type measurement taken in the same day the missense variants were tested in triplicate transcriptional assays the construct containing wild type brca amino acids to wt was used as a positive control and variants m r and y x as negative controls hek t cells were seeded in well tissue culture plates cells per well and hours later were transfected in triplicates with fugene according to the manufacturer s instructions roche the vectors containing the variants were cotransfected with pg luc which contains a firefly luciferase reporter gene driven by gal responsive binding sites and phgr tk which contains a renilla luciferase gene driven by a constitutive tk basal promoter used as an internal control cells were harvested hours after transfection and luciferase activity was measured using the dual luciferase reporter assay system promega following the manufacturer s instructions activity was plotted as a percentage of the wt brca activity thirty four of variants had already been assessed in previous articles albeit in different amino acid contexts amino acids or aa refs in only one case g a the qualitative assessment differed because it was considered of moderate effect previously and of no effect in the present study this shows the high reproducibility of the transcriptional assays align gvgd classification of brca brct variants the brct variants were analyzed with the align gvgd algorithm at three depths of alignment nine mammalian brca sequences chicken and one additional sequence either frog puffer fish or sea urchin were used for the alignment for more detail see ref a class scores were assigned for variants in which the scores from at least two of three alignments agreed definition of a validation set to compile our validation set supplementary table s table we identified variants that had been previously analyzed and classified by genetic or integrative methods integrative methods use a combination of data from different sources including cooccurrence with known deleterious mutations personal and family history of patients carrying the variant and cosegregation of the variant with disease within pedigrees to assign an overall risk score sufficient data were available to assign of the brca brct variants as associated with disease whereas of the variants were predicted to be neutral lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript table where possible the probability of pathogenicity of each of these variants was classified according to the iarc scheme supplementary table s ref we calculated specificity sensitivity and confidence interval measures for each of the structure and functional assays based on a comparison with this variant set supplementary table s a description of how the brca brct variant set was obtained is given in supplementary materials and methods results generation of a large set of brca brct missense variants many of the brca variants that have been uncovered through worldwide sequencing efforts are now deposited within the breast cancer information core database http research nhgri nih gov bic as well as the kathleen cuningham foundation consortium for research into familial breast cancer database http www kconfab org to facilitate the assessment of the structural and functional defects associated with these mutations we cloned all the currently available missense variants fig into an e coli expression vector that we used previously for our structural and functional work of the brca brct domain this vector places the brca brct coding sequence corresponding to residues under control of a t promotor using this plasmid system recombinant protein can be expressed in e coli and purified using established procedures for detailed biochemical and structural studies the t promoter can also be used for in vitro transcription translation of the recombinant protein the in vitro transcription translation system is rapid and allows the generation of multiple proteins in parallel and because the recombinant protein is specifically labeled with s methionine no purification is required all variants can be successfully made in this system regardless of their folding stability this is a further advantage over e coli expression in which we have found that only well folded variants can be expressed in soluble form and purified thus we have used in vitro transcribed translated material for our structural stability and peptidebinding assays described below we also evaluated the possibility that some of these variants could result in splicing abnormalities we used the nnsplice algorithm to predict the effects of these changes on the efficiency with the brct coding region see supplementary materials determination of the protein folding defects associated with brct missense variants we used limited proteolysis to determine the stability of the protein fold in the set of brca brct missense variants using in vitro transcribed translated material supplementary fig s the wild type protein is highly resistant to digestion by trypsin elastase or chymotrypsin indicating that the in vitro produced brct protein is stably folded on the other hand a pilot study involving of the brca brct missense variants showed that certain single amino acid substitutions in the brct domain of brca can destabilize its structure leading to increased susceptibility to trypsin mediated proteolysis the assay is sufficiently sensitive to discriminate wild type protein from the strongly destabilized variant a e as well as the moderately destabilized variant m r proteolysis of each of the missense variants was carried out at a series of trypsin concentrations and the amount of full length protein remaining was visualized by sdspage the results were compared with those of the wild type protein as well as the m r and a e variants carried out in parallel we quantified the percentages of fulllength protein remaining following digestion at different concentrations of trypsin and determined the severity of the destabilizing effects by comparing the percentage of protein remaining following digestion at \\u03bcg ml trypsin to the controls supplementary fig s a we defined highly destabilizing mutations as those with similar rates of digestion to lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript the a e control of the protein remaining after digestion with \\u03bcg ml of trypsin moderately destabilizing mutations were defined as those with a similar stability to m r between and of the protein remaining after digestion with \\u03bcg ml of trypsin finally structurally stable variants were defined as those with a stability similar to the wild type protein of the protein remaining after \\u03bcg ml of trypsin digestion based on these criteria of of the missense mutations showed high sensitivity to tryptic digestion of showed intermediate sensitivity and of showed a stability that was indistinguishable from the wild type supplementary figs s a and s we also noted that the c r mutant showed a particularly low protein expression level analysis of the phosphopeptide binding activity of brct missense variants the brca brct domain interacts with bach and several other proteins through the selective recognition of the peptide motif pser x x phe all mutations that have been strongly linked to disease using family based data such as a e and m r fail to bind specifically to pser x x phe containing peptides as shown in a pilot study that analyzed the peptide binding activities of brca brct domain variants this indicates the functional importance of the brca brct domain as a phosphopeptidebinding module in tumor suppression we examined the ability of the brct domains containing missense variants to bind peptides derived from bach using a pull down assay in which biotinylated peptide was bound to streptavidin beads we tested binding to a phosphorylated version of the peptide containing the brca target sequence pser pro thr phe or an unphosphorylated version as a control the results were compared with experiments with the wild type brca brct as well as with the peptide binding deficient variants a e and m r carried out in parallel we present two analyses of the data in the first analysis we define binding activity as the fraction of the input protein that bound to the phosphorylated peptide normalized against the binding of the wild type protein supplementary fig s b in the second analysis we define binding specificity as the ratio of the variant bound to the phosphopeptide compared with the nonphosphorylated counterpart normalized to the wildtype control supplementary fig s c classified in this way we observed a range of binding activities and specificities as observed for the protein stability assays the variants tended to cluster near either or activity with relatively few variants displaying intermediate binding activities specificities supplementary fig s thirty eight of of the variants displayed binding activities of wild type levels whereas of variants displayed binding activities of wild type the binding specificity analysis revealed of of the variants exhibited of wild type levels whereas of of the variants exhibited of wild type levels analysis of the transcriptional activation activity of brca brct variants the transcriptional activation activity of the brca brct domain was the first identified activity of this region of the protein assays for brca brct transcriptional activation have been established in human and yeast cell systems in either method the brct domain is fused to a heterologous dna binding domain that recognizes a defined regulatory sequence within the promoter of the reporter pilot studies involving subsets of the brca brct variants showed that mutations that have been associated with cancer using family based data abrogate or markedly diminish the expression of the reporter gene moreover the fact that mutations that destabilize the brct fold are strongly associated with a loss of transactivation function indicates that this assay is a monitor of the integrity of the brct domain lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript as observed for the protein stability and phosphopeptide binding assays a wide range of transcriptional activities were observed for the variant set with the observed values tending to cluster either at very low or very high activity levels supplementary figs s d and s thirty eight of variants displayed transcriptional activity levels at of wild type whereas of were found to be of wild type levels correlation of structural and functional data with family history and clinical data the use of functional data in cancer risk assessment is predicated on rigorous validation of the results against sets of variants for which cancer risks are well defined easton and colleagues have assigned risks based on genetic data for a large number of brca variants we used these data together with available genetic information for several other variants to create a validation set of variants consisting of deleterious mutations and neutral variants table we plotted the protein stability peptide binding and transcriptional activity data against the combined log likelihood ratio scores for the subset of variants analyzed by easton and colleagues fig likelihood ratio in general this analysis revealed a clear correlation between structural functional defects and cancer risk a comparison of all the genetically characterized variants with the experimental results is shown in table this comparison was used to define the sensitivities and specificities of the functional assays sensitivity ranged from for phosphopeptide binding activity to for phosphopeptide binding specificity and transcription assays specificity ranged from for proteolysis assays to for transcription assays to for binding activity assays supplementary table s although overall there is a good agreement between the assessed cancer risks and the functional activities of the variants there are several variants in which cancer risk assessment and the structure function results do not agree for example a significant number of cancer associated iarc classes and variants are not structurally destabilized as determined by the proteolysis assay each of these variants r w m r and m k is highly defective for specific phosphopeptide recognition and transcriptional activity structural studies reveal that both arg and met make critical direct contacts to the phosphopeptide and structural investigations of the m r and m k variants reveal subtle structural rearrangements that block the specificity pocket two of the cancer associated iarc class variants i s and g v were highly defective in the proteolysis and transcriptional assays however they showed significant levels of nonspecific peptide binding independent of phosphorylation it is likely that the unfolded state of these variants exposes hydrophobic residues that are responsible for these nonspecific interactions as suggested previously g e one of the cancer associated iarc class variants is defective in the proteolysis and peptidebinding assays however it shows an intermediate level of transcriptional activity the neutral iarc class variant v a exhibited the least agreement between the different assays showing significant sensitivity to proteolysis and a strong transcriptional defect while showing intermediate strong levels of phosphopeptide interaction considering the overall high specificity and sensitivity obtained in the assays supplementary table s we have defined limits to predict disease association for the uncharacterized variants based on the biochemical and transcriptional assay data table fig it is important to stress that the definition of these limits should not be used for classifying variants in terms of their pathogenicity in the absence of additional data the cancer associated iarc classes and variants in general exhibited protein stability levels of the wild type phosphopeptide binding activities of wild type phosphopeptide specificity binding activities of wild type and transcriptional activation function of wild type levels conversely the neutral variants iarc classes and in general had protein stability levels of the wild type phosphopeptide lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript binding activities of wild type phosphopeptide specificity binding activities of wild type and transcriptional activation function of wild type levels variants falling between these limits are defined as uncertain using these limits we have defined the structure and functional activities of each of the variants fig comparison of results from different assays and overall classification next we compared the results from the different assays for the entire set of variants fig in of of the variants there was complete concordance between the different assays with of giving protein stability peptide binding and transcription activities within the limits that correspond to neutral variants in iarc classes and whereas of exhibited values in the range corresponding to cancer associated iarc classes and variants for each of the assays based on these results we have classified the variants into groups based on their functional effect table although classification is straightforward for those variants in which there is a complete agreement between the different assays for the variants in which there is not a complete agreement classification is more complex the largest class of discordant results includes variants that exhibit defects in binding and transcription but little or no folding defect of variants or these variants likely directly affect the phosphopeptide binding surface and may only affect residues that are exposed to the solvent preserving the overall fold structure because phosphopeptide recognition is critical for the tumorsuppressive function of brca we have therefore included these variants in the strong functional effect class likewise severely destabilized variants that show significant levels of phosphorylation independent peptide binding but a strong transcriptional defect have also been categorized as strong functional effect table we have also described eight of the variants as of moderate functional effect in which a preponderance of the data points to a loss of function these include variants with intermediate levels of peptide binding specificity w g and a p or transcriptional activity v l g e and h n and strong defects in all of the other assays in addition we have included three additional mutants in this group that show poor agreement between the different assays but are likely to directly affect the peptide binding surface s f is predicted to impede recognition of the phosphoserine in the bound peptide whereas g d could perturb the structure of the phosphoserine binding pocket and the introduction of a negatively charged side chain in this region would be expected to further inhibit phosphopeptide recognition r l would remove a critical determinant for recognition of the phenylalanine residue at the position of the phosphopeptide and indeed the similar r w variant is pathogenic iarc class table we predict a given variant to be neutral when each of the assays gives values similar to wild type within the limits defined by comparison with the set of variants that have been characterized through family history and genetic data in addition we define a set of seven variants as low functional effect that display an intermediate activity for one of the assays and wild type levels for the other three finally there is a group of variants that we define as uncertain in general these variants show a poor level of agreement between the different assays and a significant number of assay results in the intermediate range correlation of mutant classification with predictions from align gvgd the align gvgd method http agvgd iarc fr agvgd_input php uses protein multiple sequence alignment pmsa data to provide cancer risk estimates for brca mutations in this method each missense variant is classified based on the observed amino acid lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript fluctuation at that position in the pmsa the grantham variance gv as well as a comparison of the amino acid substitution with the range of amino acids observed at that position grantham difference gd in this way seven classes of variants c c c c c c and c that represent a graded series of gv gd scores were defined variants in the c class correspond to variants that are well conserved in the pmsa and are therefore predicted to be neutral whereas the c class contains missense variants that deviate from the pmsa and likely associated with an increased cancer risk to assess the accuracy of the align gv gd method we compared align gv gd scores to the results for the structure function assays fig for the variants with no functional effect a large majority of falls into the c class three are classified as c and two are classified as c only one of them g a received a high score in align gv gd c another variant at this position g e scored as c but this variant was classified with moderate functional effect by structure function assays these results suggest substitution of a small side chain at position is tolerated but not substitution of the larger charged glu for the variants with strong functional effect of are in the c class and are spread throughout classes c to c fig b one missense variant a p was classified c but this variant shows strong defect in all assays the substitution of a proline at position likely destabilizes the linker \\u03b1 helix which may be responsible for the strong folding defect associated with this variant overall there is strong agreement between align gv gd predictions and the results of functional assays indicating that sequence alignment based predictive methods can offer a valuable measure of risk assessment for variants in which there is no direct functional data discussion here we present the first comprehensive assessment of the structural and functional consequences of brca variants corresponding to the complete set of brct missense variants detected to date in the human population coverage our protein structure assay peptide binding and transcriptional assays show a remarkable degree of agreement and give us confidence in the validity of these results correlation of this experimental data with the available clinical and family history data for a subset of these variants defines limits that can be used to tentatively predict functional effect that is likely to be reflected in the disease risk associated with these variants in this way we show that of the variants fall into the clear functional effect or moderate functional effect categories table and are likely associated with an increased cancer risk whereas fall into the no functional effect or low functional effect categories the remaining falls into the uncertain category and could perhaps represent low penetrance variants the fact that a large proportion of variants tested displayed some functional effect lends support to the idea of applying higher prior probabilities for causality to variants found in the brct domains the brca brct variant set provides a useful set of reagents for further detailed study of variants of particular interest for example variants that are moderately well folded based on proteolytic stability can often be purified and studied in more detail circular dicroism spectroscopy and differential scanning calorimetry have been used to quantitatively assess the fold stability of v l r w m k m r m t v a v f and v m although our proteolytic assay indicated that the stability of two of these variants r w and v a were indistinguishable from wild type the more sensitive thermodynamic studies of the purified forms of these variants reveal a small but significant reduction in fold stability in addition quantitative assays of phosphopeptide binding such as isothermal titration calorimetry can reveal the detailed thermodynamics of peptide brct interactions the structural rearrangements induced by three of these variants m r m k and v f have been visualized through lee et al page cancer res author manuscript available in pmc june nih pa author manuscript nih pa author manuscript nih pa author manuscript crystallographic structural analysis each of the variants affect the specificity pocket formed at the interface between the two brct repeats for m r and m k the substituted positively charged side chain adopts an orientation that fills the specificity pocket and blocks the binding of the phe side chain at the position of the target peptide intriguingly the v f substitution which occurs far from the pocket induces a series of conformational adjustments that results in the movement of met to fill the specificity pocket explaining the loss of peptide binding specificity and transcriptional activity associated with this variant nonsynonymous single nucleotide polymorphisms snp that lead to amino acid substitutions are likely to have diverse phenotypic consequences and implications for human disease although computational methods have been proposed to assess the effect of individual snps relatively little has been done at the experimental level to provide insight into the effects of snps the database of brca brca mutations provides an important first look at the kind of genetic diversity that exists in the human population another example of a large mutation database is that assembled for mutations found in p from human tumors detailed studies of the stability of the p dna binding domain indicate that like the brca brct domain this relatively unstable domain is sensitive to mutations that destabilize the protein fold in addition to other classes of mutations that do not significantly affect protein fold stability but instead modify dna binding or proteinprotein interaction surfaces thus rapid general assays of protein structure and function such as the proteolysis assay used here could potentially prove useful for the rapid high throughput assessment of the effect of snps on protein function ultimately unambiguous classification of variants as proposed by the iarc unclassified genetic variants working group will need a comprehensive integrative model to incorporate data from several sources robust integrative models have been proposed but they still lack a coherent framework to incorporate functional data a major roadblock in the development of such a model was the lack of an established series of functional assays with standardized controls validated against available genetic data the present study fills this gap and allows for the development of better predictive models although a systematic large scale classification of all brca variants into the iarc classes has not yet been done it is expected that a large number will fall into class uncertain due to the lack of genetic based data conceivably even large scale efforts may not be able to gather sufficient genetic data on individual variants because most are extremely rare thus we anticipate that the use of structural and functional information will be instrumental to move variants from the class into classes that provide a basis for better risk assessment and clinical decisions in addition the fact that variants can now also be functionally and structurally rationalized will provide an impulse for further functional dissection of the tumor suppressive activities of brca BRCA1 D1739G'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df[\"Text\"][0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:tensorflow]",
   "language": "python",
   "name": "conda-env-tensorflow-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
